{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import pysam\n",
    "import numpy as np\n",
    "import time\n",
    "from scipy.signal import savgol_filter\n",
    "import yaml \n",
    "from multiprocessing import Pool\n",
    "\n",
    "# import warnings\n",
    "# warnings.filterwarnings('error')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "#sample specific params for testing\n",
    "#ER_pos_merged:\n",
    "sample_name = 'HD45.ctDNA.WGS.FC19269447'\n",
    "bam_path = '/fh/scratch/delete90/ha_g/realigned_bams/cfDNA_deepWGS_hg38/deepWGS_fastq_to_bam_paired_snakemake/results/HD45.ctDNA.WGS.FC19269447/HD45.ctDNA.WGS.FC19269447_recalibrated.bam'\n",
    "GC_bias_path = '/fh/fast/ha_g/user/adoebley/projects/griffin_paper/GC_correction/MBC_GC_correction/results/repeat_masker.mapable.k50.Umap.hg38/GC_bias/HD45.ctDNA.WGS.FC19269447.GC_bias.txt'\n",
    "background_normalization = 'none'\n",
    "ref_seq_path = '/fh/fast/ha_g/grp/reference/GRCh38/GRCh38.fa'\n",
    "\n",
    "# #additional params for testing\n",
    "sites_yaml = 'test_sites.yaml'\n",
    "results_dir = 'results/'\n",
    "\n",
    "chrom_col = 'Chrom'\n",
    "chroms = ['chr'+str(m) for m in np.arange(1,23)]\n",
    "norm_window = [-5000, 5000] #for testing\n",
    "plot_window = [-1000, 1000]#for testing\n",
    "fragment_length = 165\n",
    "\n",
    "step = 15\n",
    "sz_range = [100, 200]\n",
    "map_q = 20\n",
    "strand_col = 'Strand'\n",
    "\n",
    "individual = 'False'\n",
    "smoothing = 'True'\n",
    "\n",
    "number_of_sites = 'none'\n",
    "sort_by = 'none'\n",
    "ascending = 'none'\n",
    "\n",
    "CPU = 4\n",
    "\n",
    "debugging = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parser = argparse.ArgumentParser()\n",
    "\n",
    "# parser.add_argument('--sample_name', help='name of sample', required=True)\n",
    "# parser.add_argument('--bam', help='bam file', required=True)\n",
    "# parser.add_argument('--GC_bias', help='GC bias info', required=True)\n",
    "# parser.add_argument('--background_normalization', help='None or local', required=True)\n",
    "# parser.add_argument('--reference_genome',help = 'path to the reference genome',required=True)\n",
    "\n",
    "# parser.add_argument('--sites_yaml', help='.bed file of sites', required=True)\n",
    "# parser.add_argument('--results_dir', help='directory for coverage_data', required=True)\n",
    "\n",
    "# parser.add_argument('--chrom_column',help='name of column containing chromosome number', default='Chrom')\n",
    "# parser.add_argument('--chroms', help='chroms to include when selecting sites', nargs='*', default=['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22'])\n",
    "# parser.add_argument('--norm_window',help='start and end of the window to be used for normalization',nargs=2, type=int, default=(-5000,5000))\n",
    "# parser.add_argument('--plot_window',help='start and end of window to be plotted',nargs=2, type=int, default=(-1000,1000))\n",
    "# parser.add_argument('--fragment_length',help='length of fragment (in bp) for which GC will be calculated, default 165',type=int, default=165)\n",
    "\n",
    "# parser.add_argument('--step',help='step size when calculating coverage', type=int, default=5)\n",
    "# parser.add_argument('--size_range',help='acceptable size range for fragments (to filter out genomic contamination)',nargs=2, type=int, default=(0,500))\n",
    "# parser.add_argument('--map_quality',help='minimum mapping quality', type=int, default=60)\n",
    "# parser.add_argument('--strand_column',help='name of column containing the strand (+ or -)', default='Strand')\n",
    "\n",
    "# parser.add_argument('--individual',help='save individual site coverage. TRUE WILL RESULT IN HUGE OUTPUT FILES. (True/False)',default='False', required = True)\n",
    "# parser.add_argument('--smoothing',help='whether to use a savgol filter to smooth sites (True/False)', required = True)\n",
    "\n",
    "# parser.add_argument('--num_sites',help='number of sites to analyze', default='NA')\n",
    "# parser.add_argument('--sort_by',help='how to select the sites to analyze', default='none')\n",
    "# parser.add_argument('--ascending',help='whether to sort in ascending or descending order when selecting sites', default='NA')\n",
    "\n",
    "# parser.add_argument('--cpu',help='cpu available for parallelizing', type = int, required = True)\n",
    "\n",
    "\n",
    "# args = parser.parse_args()\n",
    "\n",
    "\n",
    "# sample_name=args.sample_name\n",
    "# bam_path=args.bam\n",
    "# GC_bias_path=args.GC_bias\n",
    "# background_normalization = args.background_normalization\n",
    "# ref_seq_path = args.reference_genome\n",
    "\n",
    "# sites_yaml=args.sites_yaml\n",
    "# results_dir=args.results_dir\n",
    "\n",
    "# chrom_col=args.chrom_column\n",
    "# chroms = args.chroms\n",
    "# norm_window =args.norm_window\n",
    "# plot_window=args.plot_window\n",
    "# fragment_length=args.fragment_length\n",
    "\n",
    "# step=args.step\n",
    "# sz_range=args.size_range\n",
    "# map_q=args.map_quality\n",
    "# strand_col=args.strand_column\n",
    "\n",
    "# individual=args.individual\n",
    "# smoothing = args.smoothing\n",
    "\n",
    "# number_of_sites=args.num_sites\n",
    "# sort_by=args.sort_by\n",
    "# ascending=args.ascending\n",
    "\n",
    "# CPU = args.cpu\n",
    "\n",
    "# debugging = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "arguments provided:\n",
      "\tsample_name = \"HD45.ctDNA.WGS.FC19269447\"\n",
      "\tbam_path = \"/fh/scratch/delete90/ha_g/realigned_bams/cfDNA_deepWGS_hg38/deepWGS_fastq_to_bam_paired_snakemake/results/HD45.ctDNA.WGS.FC19269447/HD45.ctDNA.WGS.FC19269447_recalibrated.bam\"\n",
      "\tGC_bias_path = \"/fh/fast/ha_g/user/adoebley/projects/griffin_paper/GC_correction/MBC_GC_correction/results/repeat_masker.mapable.k50.Umap.hg38/GC_bias/HD45.ctDNA.WGS.FC19269447.GC_bias.txt\"\n",
      "\tbackground_normalization = \"none\"\n",
      "\tref_seq_path = \"/fh/fast/ha_g/grp/reference/GRCh38/GRCh38.fa\"\n",
      "\tsites_yaml = \"test_sites.yaml\"\n",
      "\tresults_dir = \"results/\"\n",
      "\tchrom_col = \"Chrom\"\n",
      "\tchroms =  ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22']\n",
      "\tnorm_window =  [-5000, 5000]\n",
      "\t#norm_window rounded to step [-4995, 4995]\n",
      "\tplot_window = [-1000, 1000]\n",
      "\t#plot_window rounded to step: [-990, 990]\n",
      "\tfragment_length = 165\n",
      "\t#fragment_length_rounded_up_to_step: 165\n",
      "\tstep = 15\n",
      "\tsz_range = [100, 200]\n",
      "\tmap_q = 20\n",
      "\tstrand_col = \"Strand\"\n",
      "\tindividual = \"False\"\n",
      "\tsmoothing = \"True\"\n",
      "\tnumber_of_sites = \"none\"\n",
      "\tsort_by = \"none\"\n",
      "\tascending = \"none\"\n",
      "\tCPU = 4\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if ascending.lower()=='false':\n",
    "    ascending=False\n",
    "elif ascending.lower()=='true':\n",
    "    ascending=True\n",
    "else:\n",
    "    ascending='none'\n",
    "\n",
    "print('\\narguments provided:')\n",
    "\n",
    "print('\\tsample_name = \"'+sample_name+'\"')\n",
    "print('\\tbam_path = \"'+bam_path+'\"')\n",
    "print('\\tGC_bias_path = \"'+GC_bias_path+'\"')\n",
    "print('\\tbackground_normalization = \"'+background_normalization+'\"')\n",
    "print('\\tref_seq_path = \"'+ref_seq_path+'\"')\n",
    "#print('\\ttumor_fraction =',tumor_fraction)\n",
    "\n",
    "print('\\tsites_yaml = \"'+sites_yaml+'\"')\n",
    "print('\\tresults_dir = \"'+results_dir+'\"')\n",
    "\n",
    "print('\\tchrom_col = \"'+chrom_col+'\"')\n",
    "print('\\tchroms = ',chroms)\n",
    "print('\\tnorm_window = ', norm_window)\n",
    "norm_window=[int(np.ceil(norm_window[0]/step)*step),int(np.floor(norm_window[1]/step)*step)] #round to the nearest step inside the window\n",
    "print('\\t#norm_window rounded to step',norm_window)\n",
    "print('\\tplot_window = '+str(plot_window))\n",
    "plot_window=[int(np.ceil(plot_window[0]/step)*step),int(np.floor(plot_window[1]/step)*step)] #round to the nearest step inside the window\n",
    "print('\\t#plot_window rounded to step:',plot_window)\n",
    "print('\\tfragment_length =',fragment_length)\n",
    "fragment_length=int(np.ceil(fragment_length/step)*step) #round fragment length to the nearest step\n",
    "print('\\t#fragment_length_rounded_up_to_step:',fragment_length)\n",
    "\n",
    "print('\\tstep =',step)\n",
    "print('\\tsz_range =',sz_range)\n",
    "print('\\tmap_q =',map_q)\n",
    "print('\\tstrand_col = \"'+strand_col+'\"')\n",
    "\n",
    "print('\\tindividual = \"'+individual+'\"')\n",
    "if smoothing.lower()=='true' or smoothing.lower()=='false':\n",
    "    print('\\tsmoothing = \"'+smoothing+'\"')\n",
    "else:\n",
    "    print('smoothing must be True or False! :',smoothing)\n",
    "    sys.exit()\n",
    "\n",
    "print('\\tnumber_of_sites = \"'+str(number_of_sites)+'\"')\n",
    "print('\\tsort_by = \"'+sort_by+'\"')\n",
    "print('\\tascending = \"'+str(ascending)+'\"')\n",
    "\n",
    "print('\\tCPU =',CPU)\n",
    "print('\\n')\n",
    "sys.stdout.flush()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define global parameters and open global files\n",
    "\n",
    "#set up global variables\n",
    "norm_columns = np.arange(norm_window[0],norm_window[1],step)\n",
    "plot_columns = np.arange(plot_window[0],plot_window[1],step)\n",
    "\n",
    "# #import the site files\n",
    "with open(sites_yaml,'r') as f:\n",
    "    site_files = yaml.safe_load(f)\n",
    "site_files = site_files['site_files']\n",
    "site_files = [[key,site_files[key]] for key in site_files.keys()]\n",
    "to_do_list = site_files\n",
    "\n",
    "########################################\n",
    "#GET GC BIAS\n",
    "########################################\n",
    "#open the GC_bias file \n",
    "GC_bias = pd.read_csv(GC_bias_path, sep='\\t')\n",
    "\n",
    "#get rid of extremely low GC bias values\n",
    "#these fragments will now be excluded \n",
    "#these fragments are extremely rare so it is difficult to get a good estimate of GC bias\n",
    "GC_bias['smoothed_GC_bias'] = np.where(GC_bias['smoothed_GC_bias']<0.05,np.nan,GC_bias['smoothed_GC_bias'])\n",
    "\n",
    "GC_bias = GC_bias[['length','num_GC','smoothed_GC_bias']]\n",
    "GC_bias = GC_bias.set_index(['num_GC','length']).unstack()\n",
    "\n",
    "#convert to a dictionary\n",
    "GC_bias = GC_bias.to_dict()\n",
    "\n",
    "#get rid of values where the num_GC is greater than the length (included due to the way I made the dict)\n",
    "GC_bias2 = {}\n",
    "for key in GC_bias.keys():\n",
    "    length = key[1]\n",
    "    GC_bias2[length] = {}\n",
    "    for num_GC in range(0,length+1):\n",
    "        bias = GC_bias[key][num_GC]\n",
    "        GC_bias2[length][num_GC]=bias\n",
    "GC_bias = GC_bias2 \n",
    "del(GC_bias2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "step_1_import_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def step_1_import_sites(sites_file,site_name): #number of sites needed so it can be changed to an int\n",
    "    all_sites=pd.read_csv(sites_file,sep='\\t')\n",
    "    print(site_name,'total sites',len(all_sites))\n",
    "    \n",
    "    #throw out sites that aren't on the selected chroms\n",
    "    all_sites = all_sites[all_sites[chrom_col].isin(chroms)]\n",
    "    print(site_name,'sites on selected chromosomes',len(all_sites))\n",
    "    \n",
    "    #select the sites to use if specified\n",
    "    if sort_by.lower()=='none': #if using all sites \n",
    "        print(site_name,'processing all '+str(len(all_sites))+' sites')\n",
    "    \n",
    "    else: #othewise sort by the specified column\n",
    "        print(site_name,'sorting by',sort_by,'and selecting the top',number_of_sites,',ascending:',ascending)\n",
    "        print(site_name,sort_by,'initial range: ',min(all_sites[sort_by]),'to',max(all_sites[sort_by]))\n",
    "        all_sites=all_sites.sort_values(by=sort_by,ascending=ascending).reset_index(drop=True)#sort and reset index\n",
    "        all_sites=all_sites.iloc[0:int(number_of_sites)]\n",
    "        print(site_name,sort_by,'range after sorting: ',min(all_sites[sort_by]),'to',max(all_sites[sort_by]))\n",
    "\n",
    "    #add a site_name column\n",
    "    all_sites['site_name']=site_name\n",
    "    #add a sample column\n",
    "    all_sites['sample']=sample_name\n",
    "    #add background normalizatino\n",
    "    all_sites['background_normalization']=background_normalization\n",
    "    \n",
    "    #split the list of all sites into groups for quicker processing\n",
    "    max_chunk_size=500\n",
    "    n_chunks=np.ceil(len(all_sites)/max_chunk_size)\n",
    "    site_groups=np.array_split(all_sites.index.values,n_chunks)\n",
    "    \n",
    "    return(all_sites, site_groups)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "step_2_collect_coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def collect_fragments(sites,window_start,window_end,direction):\n",
    "    #open the bam file for each pool worker (otherwise individual pool workers can close it)\n",
    "    bam_file = pysam.AlignmentFile(bam_path)\n",
    "    \n",
    "    #open the ref seq\n",
    "    ref_seq=pysam.FastaFile(ref_seq_path)\n",
    "\n",
    "    #extend the window by half the max fragment length in each direction\n",
    "    max_adjustment = np.int(np.ceil(sz_range[1]/2))\n",
    "    adjusted_start = window_start-max_adjustment\n",
    "    adjusted_end = window_end+max_adjustment\n",
    "    \n",
    "    window_columns = np.arange(window_start,window_end,step)\n",
    "    \n",
    "    #make sure Chrom is a string\n",
    "    sites[chrom_col]=sites[chrom_col].astype(str)\n",
    "            \n",
    "    cov_pd={} #set up dictionary to hold data\n",
    "    GC_cov_pd = {} #set up a dictionary to hold GC corrected data\n",
    "\n",
    "\n",
    "    for endpoint in ['start','end']:\n",
    "        cov_pd[endpoint]=pd.DataFrame(columns = window_columns)\n",
    "        GC_cov_pd[endpoint] = pd.DataFrame(columns = window_columns)\n",
    "        \n",
    "    #workaround for NCBI style formatted bams (1,2,3 etc) with UCSC sites\n",
    "    if len(sites)>0:\n",
    "        test_chrom=sites.iloc[0][chrom_col]\n",
    "        try:\n",
    "            bam_file.get_reference_length(test_chrom)\n",
    "            NCBI=False\n",
    "        except:\n",
    "            bam_file.get_reference_length(test_chrom.split('chr')[-1])\n",
    "            NCBI=True\n",
    "     \n",
    "    #run analysis on each site\n",
    "    for i in range(len(sites)): #for each location in the genome\n",
    "        if i%100==0: \n",
    "            print (sites.iloc[0]['site_name'],i,time.time()-start_time)\n",
    "            sys.stdout.flush()\n",
    "\n",
    "        #############\n",
    "        #make dicts to hold output for this individual site\n",
    "        #############\n",
    "        #for now, catch all reads that start or end within a fragment length of the window\n",
    "        cov_dict = {}\n",
    "        GC_cov_dict = {}\n",
    "        for item in ['start','end']:\n",
    "            cov_dict[item]={m:0 for m in range(adjusted_start,adjusted_end)} \n",
    "            GC_cov_dict[item]={m:0 for m in range(adjusted_start,adjusted_end)} \n",
    "                \n",
    "        ####################\n",
    "        #fetch reads\n",
    "        ####################\n",
    "        #identify the analysis window for that site\n",
    "        chrom = sites.iloc[i][chrom_col]\n",
    "        position = sites.iloc[i]['position']\n",
    "        \n",
    "        if NCBI:\n",
    "            chrom=chrom.split('chr')[-1]\n",
    "                    \n",
    "        #these regions have been filtered so they should all be fetchable\n",
    "        fetched=bam_file.fetch(contig=chrom, start=adjusted_start+position, stop=adjusted_end+position) #fetch reads that map to the region of interest\n",
    "        \n",
    "        ########################\n",
    "        #count coverage\n",
    "        ########################\n",
    "        for read in fetched:\n",
    "            #filter out reads\n",
    "            if abs(read.template_length)>=sz_range[0] and abs(read.template_length)<=sz_range[1] \\\n",
    "            and read.is_paired==True and read.mapping_quality>=map_q and read.is_duplicate==False and read.is_qcfail==False:\n",
    "                read_start=read.reference_start-position #read start (with respect to the position of the current region of interest)\n",
    "                \n",
    "                #find the place where the fragment starts or ends with respect to the window\n",
    "                if read.is_reverse==False and read.template_length>0:\n",
    "                    fragment_start = read_start\n",
    "                    fragment_end = read_start+read.template_length\n",
    "                    read_type='start'\n",
    "                    \n",
    "                    #adjusted_loc = fragment_start+int(np.ceil(fragment_length/2)) #where to count the coverage\n",
    "                    midpoint = int(np.floor((fragment_start+fragment_end)/2))\n",
    "                                        \n",
    "                elif read.is_reverse==True and read.template_length<0:\n",
    "                    read_len=read.reference_length #this is the read length only (for adjusting the start position)\n",
    "                    fragment_start = read_start+read_len+read.template_length\n",
    "                    fragment_end = read_start+read_len\n",
    "                    read_type='end'\n",
    "                    \n",
    "                    #adjusted_loc = fragment_end-int(np.ceil(fragment_length/2)) #where to count the coverage\n",
    "                    midpoint = int(np.floor((fragment_start+fragment_end)/2))\n",
    "                           \n",
    "                else:\n",
    "                    continue\n",
    "                    \n",
    "                #get the fragment seq for GC content\n",
    "                #seq_string = (ref_seq.fetch(chrom,fragment_start+position,fragment_end+position)).upper() #for printing\n",
    "                \n",
    "                fragment_seq = (ref_seq.fetch(chrom,fragment_start+position,fragment_end+position)).upper()\n",
    "                fragment_seq = list(fragment_seq.replace('T','0').replace('A','0').replace('C','1').replace('G','1').replace('N',str(np.random.randint(0,2))))\n",
    "                fragment_seq = [int(m) for m in fragment_seq]\n",
    "                \n",
    "                ##check work\n",
    "                ############\n",
    "                #print(read_type,read.query_sequence)\n",
    "                #if read.is_reverse == False:\n",
    "                #    print(read_type,seq_string[0:read.reference_length])\n",
    "                #elif read.is_reverse == True:\n",
    "                #    print(read_type,seq_string[-read.reference_length:])\n",
    "                #print(sum(fragment_seq),len(fragment_seq),read.template_length)\n",
    "                #print('\\n')\n",
    "                ###########\n",
    "                \n",
    "                #check that the site is in the window          \n",
    "                if midpoint>=adjusted_start and midpoint<adjusted_end:\n",
    "                    #count the fragment\n",
    "                    cov_dict[read_type][midpoint]+=1\n",
    "                    \n",
    "                    ##get the GC bias\n",
    "                    read_GC_content = sum(fragment_seq)\n",
    "                    read_GC_bias = GC_bias[abs(read.template_length)][read_GC_content]\n",
    "                    \n",
    "                    #count the fragment weighted by GC bias\n",
    "                    if not np.isnan(read_GC_bias):\n",
    "                        GC_cov_dict[read_type][midpoint]+=(1/read_GC_bias)\n",
    "   \n",
    "                else: #if fragment doesn't fully overlap\n",
    "                    continue\n",
    "            \n",
    "                del(read,midpoint,read_type)\n",
    "            \n",
    "        ###########################      \n",
    "        #after counting all reads at that site\n",
    "        ###########################\n",
    "        for endpoint in ['start','end']:\n",
    "            cov_list = np.array([cov_dict[endpoint][m] for m in range(window_start,window_end)])\n",
    "            cov_list = np.sum(cov_list.reshape(-1, step), axis=1) #take the sum of every step\n",
    "            cov_list = pd.Series(cov_list, index = window_columns, name=sites.iloc[i].name) #convert to pandas series\n",
    "            cov_pd[endpoint]=cov_pd[endpoint].append(cov_list)\n",
    "            GC_cov_list=np.array([GC_cov_dict[endpoint][m] for m in range(window_start,window_end)])\n",
    "            GC_cov_list=np.sum(GC_cov_list.reshape(-1, step), axis=1) #take the sum of every 5(step) bp\n",
    "            GC_cov_list = pd.Series(GC_cov_list, index = window_columns, name=sites.iloc[i].name) #convert to pandas series\n",
    "            GC_cov_pd[endpoint]=GC_cov_pd[endpoint].append(GC_cov_list)\n",
    "\n",
    "    #put into the same format as the output data\n",
    "    out_dict = {}\n",
    "    for data_type in ['reads','GC_corr']:\n",
    "        out_dict[data_type]={}\n",
    "        for endpoint in ['start','end']:\n",
    "            out_dict[data_type][endpoint]={}    \n",
    "    \n",
    "    for endpoint in ['start','end']:\n",
    "        out_dict['reads'][endpoint][direction] = cov_pd[endpoint]\n",
    "        out_dict['GC_corr'][endpoint][direction] = GC_cov_pd[endpoint]\n",
    "        \n",
    "    return(out_dict)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "run full analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_full_analysis(input_items):\n",
    "    site_name,site_file = input_items\n",
    "\n",
    "    #############################################\n",
    "    #check whether the output files already exist\n",
    "    #############################################\n",
    "    current_out_file=results_dir+'/'+sample_name+'.'+site_name+'.midpoint_coverage.txt'\n",
    "    if os.path.exists(current_out_file):\n",
    "        print('done',site_name)\n",
    "        return('done')\n",
    "    else: #if any file is incomplete\n",
    "        print('analyzing:',site_name)\n",
    "\n",
    "    #make any necessary directorys for output\n",
    "#     try: \n",
    "#         os.mkdir(results_dir+'/coverage/'+site_name)\n",
    "#         print('output directory created:',results_dir+'/coverage/'+site_name)\n",
    "#     except:\n",
    "#         print('output directory already exists:',results_dir+'/coverage/'+site_name)\n",
    "#         pass\n",
    "\n",
    "    ############\n",
    "    #set up#####\n",
    "    ############\n",
    "    all_sites,site_groups = step_1_import_sites(site_file,site_name)\n",
    "\n",
    "\n",
    "    #set up structure for output\n",
    "    out_data = {}\n",
    "    for data_type in ['reads','GC_corr']:\n",
    "        out_data[data_type]={}\n",
    "        for endpoint in ['start','end']:\n",
    "            out_data[data_type][endpoint]={}\n",
    "            for direction in ['fw','rv']:#site direction\n",
    "                    out_data[data_type][endpoint][direction]=pd.DataFrame()\n",
    "\n",
    "    ################################\n",
    "    #process each group of sites####\n",
    "    ################################ \n",
    "    for group in site_groups:           \n",
    "        sites=all_sites.loc[group].copy() #get the sites at the specified indexes\n",
    "\n",
    "        print(site_name,'processing sites:',sites.index.values[0],'to',sites.index.values[-1])\n",
    "        sys.stdout.flush()\n",
    "\n",
    "        #if site direction is specified, split fw and rv sites\n",
    "        if strand_col in sites.columns:\n",
    "            fw_sites=sites[~(sites[strand_col]=='-')].copy() #include non-specified sites\n",
    "            rv_sites=sites[sites[strand_col]=='-'].copy()\n",
    "            print(site_name,'fw,rv',len(fw_sites),len(rv_sites))\n",
    "        else: #if direction isn't specified, put all sites into fw sites\n",
    "            fw_sites = sites.copy()\n",
    "            rv_sites = pd.DataFrame(columns = sites.columns)\n",
    "\n",
    "        current_fw_data = collect_fragments(fw_sites,norm_window[0],norm_window[1],'fw')\n",
    "        current_rv_data = collect_fragments(rv_sites,norm_window[1]*-1,norm_window[0]*-1,'rv')\n",
    "\n",
    "        #merge current fw data into data frame\n",
    "        for data_type in ['reads','GC_corr']:\n",
    "            for endpoint in ['start','end']:\n",
    "                out_data[data_type][endpoint]['fw'] = out_data[data_type][endpoint]['fw'].append(current_fw_data[data_type][endpoint]['fw'])\n",
    "\n",
    "        #merge current rv data into data frame\n",
    "        for data_type in ['reads','GC_corr']:\n",
    "            for endpoint in ['start','end']:\n",
    "                out_data[data_type][endpoint]['rv'] = out_data[data_type][endpoint]['rv'].append(current_rv_data[data_type][endpoint]['rv'])\n",
    "\n",
    "    #flip the reverse sites\n",
    "    print(site_name,'flipping reverse sites')\n",
    "    sys.stdout.flush()\n",
    "\n",
    "    for data_type in ['reads','GC_corr']:\n",
    "        for endpoint in ['start','end']:\n",
    "            new_columns = (out_data[data_type][endpoint]['rv'].columns*-1)-step\n",
    "            out_data[data_type][endpoint]['rv'].columns = new_columns\n",
    "            out_data[data_type][endpoint]['rv']=out_data[data_type][endpoint]['rv'][np.sort(new_columns)]\n",
    "        out_data[data_type]['start']['rv'],out_data[data_type]['end']['rv'] = out_data[data_type]['end']['rv'],out_data[data_type]['start']['rv']\n",
    "\n",
    "    #merge fw and rv sites into a single data frame\n",
    "    for data_type in ['reads','GC_corr']:\n",
    "        for endpoint in ['start','end']:\n",
    "            out_data[data_type][endpoint] = out_data[data_type][endpoint]['fw'].append(out_data[data_type][endpoint]['rv']) \n",
    "            #sort them back into the original order\n",
    "            out_data[data_type][endpoint] = out_data[data_type][endpoint].sort_index()\n",
    "\n",
    "    #get the total reads per site\n",
    "    start_total_reads = out_data['reads']['start'].sum(axis=1)\n",
    "    end_total_reads = out_data['reads']['end'].sum(axis=1)\n",
    "    GC_start_total_reads = out_data['GC_corr']['start'].sum(axis=1)\n",
    "    GC_end_total_reads = out_data['GC_corr']['end'].sum(axis=1)\n",
    "\n",
    "    #add to metadata data frame\n",
    "    all_sites['total_read_starts'] = start_total_reads\n",
    "    all_sites['total_read_ends'] = end_total_reads \n",
    "    all_sites['GC_corrected_total_read_starts'] = GC_start_total_reads\n",
    "    all_sites['GC_corrected_total_read_ends'] = GC_end_total_reads\n",
    "    \n",
    "    #normalize the average site to 1\n",
    "    for data_type in ['reads','GC_corr']:\n",
    "        for endpoint in ['start','end']:\n",
    "            #i'm not sure why this if statement is in here, not going to delete it quite yet\n",
    "            #if len(out_data[data_type][endpoint])>0:\n",
    "            #take the means (ignore nans)\n",
    "            mean_value = np.nanmean(out_data[data_type][endpoint].astype(float).values) \n",
    "            #normalize to 1 for an average site \n",
    "            out_data[data_type][endpoint] = out_data[data_type][endpoint]/mean_value\n",
    "\n",
    "    #normalize individual sites to 1 if background_normalization == 'local'\n",
    "    if background_normalization == 'local':\n",
    "        print(site_name,'normalizing individual sites to one')\n",
    "        for data_type in ['reads','GC_corr']:\n",
    "            for endpoint in ['start','end']:\n",
    "                mean_data = out_data[data_type][endpoint].values.mean(axis=1,keepdims=True)\n",
    "                #replace zero with nan\n",
    "                mean_data = np.where(mean_data==0,np.nan,mean_data)\n",
    "                out_data[data_type][endpoint] = out_data[data_type][endpoint]/mean_data\n",
    "    \n",
    "\n",
    "    #if not saving individual sites, take the mean\n",
    "    if individual.lower()=='false':\n",
    "        print(site_name,'taking means')\n",
    "        for data_type in ['reads','GC_corr']:\n",
    "            for endpoint in ['start','end']:\n",
    "                #not sure why 'astype' is needed but it seems to be required    \n",
    "                #number of sites will be the same for all data types and endpoints        \n",
    "                number_of_sites_used = len(out_data[data_type][endpoint][~(np.isnan(out_data[data_type][endpoint].values.mean(axis=1).astype(float)))])\n",
    "                            \n",
    "                out_data[data_type][endpoint] = out_data[data_type][endpoint].mean(axis=0)\n",
    "                #convert back to dataframe for further processing\n",
    "                out_data[data_type][endpoint] = pd.DataFrame(out_data[data_type][endpoint]).T\n",
    "        print(site_name,'mean of',number_of_sites_used, 'sites')\n",
    "\n",
    "    #################\n",
    "    #smooth data\n",
    "    #################\n",
    "    if smoothing.lower()=='true':\n",
    "        #savgol window should be approx one fragment length but it must be odd\n",
    "        savgol_window=np.floor(fragment_length/step)\n",
    "        if savgol_window%2==0:\n",
    "            savgol_window=savgol_window+1\n",
    "        savgol_window=int(savgol_window)\n",
    "\n",
    "        print(site_name,'smoothing')\n",
    "        sys.stdout.flush()\n",
    "\n",
    "        for data_type in ['reads','GC_corr']:\n",
    "            for endpoint in ['start','end']:\n",
    "                out_data[data_type][endpoint][norm_columns]=savgol_filter(out_data[data_type][endpoint][norm_columns], savgol_window, 3)\n",
    "    #################\n",
    "\n",
    "    #keep only the plotting columns\n",
    "    for data_type in ['reads','GC_corr']:\n",
    "        out_data[data_type]['start'] = out_data[data_type]['start'][plot_columns]\n",
    "        out_data[data_type]['end'] = out_data[data_type]['end'][plot_columns]\n",
    "\n",
    "\n",
    "    #if taking the mean of all sites, get metadata for a mean site\n",
    "    if individual.lower()=='false':\n",
    "        print(site_name,'taking metadata means')\n",
    "        metadata = pd.DataFrame(all_sites.iloc[0][['site_name','sample','background_normalization']]).T\n",
    "        metadata['total_read_starts'] = np.mean(all_sites['total_read_starts'])\n",
    "        metadata['total_read_ends'] = np.mean(all_sites['total_read_ends'])\n",
    "        metadata['GC_corrected_total_read_starts'] = np.mean(all_sites['GC_corrected_total_read_starts'])\n",
    "        metadata['GC_corrected_total_read_ends'] = np.mean(all_sites['GC_corrected_total_read_ends'])\n",
    "        metadata['number_of_sites'] = number_of_sites_used\n",
    "        all_sites = metadata\n",
    "\n",
    "    #merge all the different dataframes\n",
    "    final_data = pd.DataFrame()\n",
    "    for data_type in ['reads','GC_corr']:\n",
    "        for endpoint in ['start','end']:\n",
    "            current_data = all_sites.merge(out_data[data_type][endpoint], left_index=True ,right_index=True)\n",
    "\n",
    "            if data_type == 'GC_corr':\n",
    "                current_data['GC_correction']='GC_corrected'\n",
    "            elif data_type == 'reads':\n",
    "                current_data['GC_correction']='none'\n",
    "            current_data['endpoint']=endpoint\n",
    "\n",
    "            final_data = final_data.append(current_data,ignore_index=True, sort=True)\n",
    "\n",
    "\n",
    "    #rearrange the columns back into a logical order\n",
    "    metadata_columns = list(all_sites.columns) + ['GC_correction','endpoint']\n",
    "\n",
    "    final_data = final_data[metadata_columns+list(plot_columns)]\n",
    "\n",
    "    #start and end traces should be nearly the same (because they are being used to calculate midpoints)\n",
    "    #keep only start for export\n",
    "    final_data = final_data[final_data['endpoint']=='start']\n",
    "    final_data['endpoint']='midpoint'\n",
    "    \n",
    "    #export\n",
    "    print(site_name,'exporting')\n",
    "    sys.stdout.flush()\n",
    "    final_data.to_csv(current_out_file,sep='\\t',float_format='%.5f', index=False)\n",
    "    \n",
    "    print('done',site_name)\n",
    "    sys.stdout.flush()\n",
    "    \n",
    "    if debugging == True:\n",
    "        print('debugging!')\n",
    "        return(final_data)\n",
    "    elif debugging == False:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "analyzing: LYL1.hg38\n",
      "LYL1.hg38 total sites 10000\n",
      "LYL1.hg38 sites on selected chromosomes 10000\n",
      "LYL1.hg38 processing all 10000 sites\n",
      "LYL1.hg38 processing sites: 0 to 499\n",
      "LYL1.hg38 0 0.31997156143188477\n",
      "LYL1.hg38 100 30.56232190132141\n",
      "LYL1.hg38 200 56.18216872215271\n",
      "LYL1.hg38 300 86.20021605491638\n",
      "LYL1.hg38 400 113.16256380081177\n",
      "LYL1.hg38 processing sites: 500 to 999\n",
      "LYL1.hg38 0 143.1016345024109\n",
      "LYL1.hg38 100 172.58060574531555\n",
      "LYL1.hg38 200 199.00035333633423\n",
      "LYL1.hg38 300 227.70034074783325\n",
      "LYL1.hg38 400 256.42046070098877\n",
      "LYL1.hg38 processing sites: 1000 to 1499\n",
      "LYL1.hg38 0 297.39291644096375\n",
      "LYL1.hg38 100 325.39811611175537\n",
      "LYL1.hg38 200 353.70874094963074\n",
      "LYL1.hg38 300 378.06776332855225\n",
      "LYL1.hg38 400 405.900897026062\n",
      "LYL1.hg38 processing sites: 1500 to 1999\n",
      "LYL1.hg38 0 430.5356812477112\n",
      "LYL1.hg38 100 454.8881573677063\n",
      "LYL1.hg38 200 480.9410512447357\n",
      "LYL1.hg38 300 507.8558406829834\n",
      "LYL1.hg38 400 530.5383937358856\n",
      "LYL1.hg38 processing sites: 2000 to 2499\n",
      "LYL1.hg38 0 555.1099934577942\n",
      "LYL1.hg38 100 577.7419493198395\n",
      "LYL1.hg38 200 603.305755853653\n",
      "LYL1.hg38 300 630.299275636673\n",
      "LYL1.hg38 400 657.1209836006165\n",
      "LYL1.hg38 processing sites: 2500 to 2999\n",
      "LYL1.hg38 0 683.0734536647797\n",
      "LYL1.hg38 100 707.9675407409668\n",
      "LYL1.hg38 200 731.5693147182465\n",
      "LYL1.hg38 300 755.2752752304077\n",
      "LYL1.hg38 400 782.7695503234863\n",
      "LYL1.hg38 processing sites: 3000 to 3499\n",
      "LYL1.hg38 0 807.3206856250763\n",
      "LYL1.hg38 100 832.9741888046265\n",
      "LYL1.hg38 200 857.3450927734375\n",
      "LYL1.hg38 300 882.1114721298218\n",
      "LYL1.hg38 400 904.8486337661743\n",
      "LYL1.hg38 processing sites: 3500 to 3999\n",
      "LYL1.hg38 0 930.9596545696259\n",
      "LYL1.hg38 100 957.6816818714142\n",
      "LYL1.hg38 200 982.3532466888428\n",
      "LYL1.hg38 300 1007.8932137489319\n",
      "LYL1.hg38 400 1032.8743298053741\n",
      "LYL1.hg38 processing sites: 4000 to 4499\n",
      "LYL1.hg38 0 1058.4189369678497\n",
      "LYL1.hg38 100 1082.8711006641388\n",
      "LYL1.hg38 200 1106.750653743744\n",
      "LYL1.hg38 300 1133.4081966876984\n",
      "LYL1.hg38 400 1160.642410993576\n",
      "LYL1.hg38 processing sites: 4500 to 4999\n",
      "LYL1.hg38 0 1187.97807097435\n",
      "LYL1.hg38 100 1214.7278962135315\n",
      "LYL1.hg38 200 1241.2187588214874\n",
      "LYL1.hg38 300 1265.972862958908\n",
      "LYL1.hg38 400 1292.964379787445\n",
      "LYL1.hg38 processing sites: 5000 to 5499\n",
      "LYL1.hg38 0 1321.350517988205\n",
      "LYL1.hg38 100 1348.094918012619\n",
      "LYL1.hg38 200 1373.7991299629211\n",
      "LYL1.hg38 300 1400.017383813858\n",
      "LYL1.hg38 400 1428.144066810608\n",
      "LYL1.hg38 processing sites: 5500 to 5999\n",
      "LYL1.hg38 0 1455.751386165619\n",
      "LYL1.hg38 100 1487.6090016365051\n",
      "LYL1.hg38 200 1513.9269032478333\n",
      "LYL1.hg38 300 1537.1634647846222\n",
      "LYL1.hg38 400 1563.7600226402283\n",
      "LYL1.hg38 processing sites: 6000 to 6499\n",
      "LYL1.hg38 0 1588.135318994522\n",
      "LYL1.hg38 100 1611.9317035675049\n",
      "LYL1.hg38 200 1635.9273283481598\n",
      "LYL1.hg38 300 1660.923615694046\n",
      "LYL1.hg38 400 1683.6807866096497\n",
      "LYL1.hg38 processing sites: 6500 to 6999\n",
      "LYL1.hg38 0 1708.8930020332336\n",
      "LYL1.hg38 100 1731.960237979889\n",
      "LYL1.hg38 200 1756.9841840267181\n",
      "LYL1.hg38 300 1781.3950307369232\n",
      "LYL1.hg38 400 1803.7083489894867\n",
      "LYL1.hg38 processing sites: 7000 to 7499\n",
      "LYL1.hg38 0 1827.9274690151215\n",
      "LYL1.hg38 100 1855.0479617118835\n",
      "LYL1.hg38 200 1880.648756980896\n",
      "LYL1.hg38 300 1909.2091717720032\n",
      "LYL1.hg38 400 1936.8926403522491\n",
      "LYL1.hg38 processing sites: 7500 to 7999\n",
      "LYL1.hg38 0 1964.9303991794586\n",
      "LYL1.hg38 100 1992.8980178833008\n",
      "LYL1.hg38 200 2018.473551273346\n",
      "LYL1.hg38 300 2043.3151862621307\n",
      "LYL1.hg38 400 2067.6416914463043\n",
      "LYL1.hg38 processing sites: 8000 to 8499\n",
      "LYL1.hg38 0 2094.5711007118225\n",
      "LYL1.hg38 100 2119.1973552703857\n",
      "LYL1.hg38 200 2143.1061582565308\n",
      "LYL1.hg38 300 2168.7085297107697\n",
      "LYL1.hg38 400 2195.6224949359894\n",
      "LYL1.hg38 processing sites: 8500 to 8999\n",
      "LYL1.hg38 0 2222.6813402175903\n",
      "LYL1.hg38 100 2247.444878578186\n",
      "LYL1.hg38 200 2273.701576948166\n",
      "LYL1.hg38 300 2300.9785640239716\n",
      "LYL1.hg38 400 2324.2997913360596\n",
      "LYL1.hg38 processing sites: 9000 to 9499\n",
      "LYL1.hg38 0 2351.0883688926697\n",
      "LYL1.hg38 100 2375.4737763404846\n",
      "LYL1.hg38 200 2403.92818069458\n",
      "LYL1.hg38 300 2431.621792078018\n",
      "LYL1.hg38 400 2459.5115745067596\n",
      "LYL1.hg38 processing sites: 9500 to 9999\n",
      "LYL1.hg38 0 2485.8142890930176\n",
      "LYL1.hg38 100 2513.4119284152985\n",
      "LYL1.hg38 200 2539.913489818573\n",
      "LYL1.hg38 300 2567.675642967224\n",
      "LYL1.hg38 400 2591.2006385326385\n",
      "LYL1.hg38 flipping reverse sites\n",
      "LYL1.hg38 taking means\n",
      "LYL1.hg38 mean of 10000 sites\n",
      "LYL1.hg38 smoothing\n",
      "LYL1.hg38 taking metadata means\n",
      "LYL1.hg38 exporting\n",
      "done LYL1.hg38\n",
      "debugging!\n",
      "analyzing: CTCF.hg38\n",
      "CTCF.hg38 total sites 10000\n",
      "CTCF.hg38 sites on selected chromosomes 10000\n",
      "CTCF.hg38 processing all 10000 sites\n",
      "CTCF.hg38 processing sites: 0 to 499\n",
      "CTCF.hg38 0 2623.5683319568634\n",
      "CTCF.hg38 100 2651.6606707572937\n",
      "CTCF.hg38 200 2682.789319753647\n",
      "CTCF.hg38 300 2710.2919461727142\n",
      "CTCF.hg38 400 2739.3559868335724\n",
      "CTCF.hg38 processing sites: 500 to 999\n",
      "CTCF.hg38 0 2769.2007455825806\n",
      "CTCF.hg38 100 2797.097110748291\n",
      "CTCF.hg38 200 2825.0632824897766\n",
      "CTCF.hg38 300 2852.099862098694\n",
      "CTCF.hg38 400 2878.922298669815\n",
      "CTCF.hg38 processing sites: 1000 to 1499\n",
      "CTCF.hg38 0 2907.9652655124664\n",
      "CTCF.hg38 100 2934.6429903507233\n",
      "CTCF.hg38 200 2961.6452989578247\n",
      "CTCF.hg38 300 2988.3907992839813\n",
      "CTCF.hg38 400 3019.239509820938\n",
      "CTCF.hg38 processing sites: 1500 to 1999\n",
      "CTCF.hg38 0 3049.581591129303\n",
      "CTCF.hg38 100 3075.0670626163483\n",
      "CTCF.hg38 200 3101.3828580379486\n",
      "CTCF.hg38 300 3126.900200366974\n",
      "CTCF.hg38 400 3153.676864385605\n",
      "CTCF.hg38 processing sites: 2000 to 2499\n",
      "CTCF.hg38 0 3181.229195833206\n",
      "CTCF.hg38 100 3208.8052763938904\n",
      "CTCF.hg38 200 3233.493528366089\n",
      "CTCF.hg38 300 3260.6547944545746\n",
      "CTCF.hg38 400 3288.397836446762\n",
      "CTCF.hg38 processing sites: 2500 to 2999\n",
      "CTCF.hg38 0 3316.580350637436\n",
      "CTCF.hg38 100 3345.426014661789\n",
      "CTCF.hg38 200 3374.7743322849274\n",
      "CTCF.hg38 300 3401.702858686447\n",
      "CTCF.hg38 400 3429.527601003647\n",
      "CTCF.hg38 processing sites: 3000 to 3499\n",
      "CTCF.hg38 0 3458.526039123535\n",
      "CTCF.hg38 100 3487.252273082733\n",
      "CTCF.hg38 200 3515.992959499359\n",
      "CTCF.hg38 300 3542.638078689575\n",
      "CTCF.hg38 400 3569.0444571971893\n",
      "CTCF.hg38 processing sites: 3500 to 3999\n",
      "CTCF.hg38 0 3598.601229429245\n",
      "CTCF.hg38 100 3625.5039341449738\n",
      "CTCF.hg38 200 3651.751315355301\n",
      "CTCF.hg38 300 3678.6489317417145\n",
      "CTCF.hg38 400 3707.1923401355743\n",
      "CTCF.hg38 processing sites: 4000 to 4499\n",
      "CTCF.hg38 0 3733.709068775177\n",
      "CTCF.hg38 100 3758.9833614826202\n",
      "CTCF.hg38 200 3786.46378159523\n",
      "CTCF.hg38 300 3813.604732990265\n",
      "CTCF.hg38 400 3842.288489818573\n",
      "CTCF.hg38 processing sites: 4500 to 4999\n",
      "CTCF.hg38 0 3871.3034806251526\n",
      "CTCF.hg38 100 3896.268304824829\n",
      "CTCF.hg38 200 3923.882091999054\n",
      "CTCF.hg38 300 3951.7096729278564\n",
      "CTCF.hg38 400 3978.1640996932983\n",
      "CTCF.hg38 processing sites: 5000 to 5499\n",
      "CTCF.hg38 0 4004.12770819664\n",
      "CTCF.hg38 100 4030.229271888733\n",
      "CTCF.hg38 200 4058.731621980667\n",
      "CTCF.hg38 300 4085.6431062221527\n",
      "CTCF.hg38 400 4111.438278198242\n",
      "CTCF.hg38 processing sites: 5500 to 5999\n",
      "CTCF.hg38 0 4138.786936998367\n",
      "CTCF.hg38 100 4165.482174396515\n",
      "CTCF.hg38 200 4193.391426086426\n",
      "CTCF.hg38 300 4223.310175895691\n",
      "CTCF.hg38 400 4247.447261333466\n",
      "CTCF.hg38 processing sites: 6000 to 6499\n",
      "CTCF.hg38 0 4276.43789768219\n",
      "CTCF.hg38 100 4303.342742443085\n",
      "CTCF.hg38 200 4334.343217611313\n",
      "CTCF.hg38 300 4361.496519088745\n",
      "CTCF.hg38 400 4388.423057317734\n",
      "CTCF.hg38 processing sites: 6500 to 6999\n",
      "CTCF.hg38 0 4414.764139175415\n",
      "CTCF.hg38 100 4442.686213731766\n",
      "CTCF.hg38 200 4469.572202205658\n",
      "CTCF.hg38 300 4493.971326112747\n",
      "CTCF.hg38 400 4520.520857810974\n",
      "CTCF.hg38 processing sites: 7000 to 7499\n",
      "CTCF.hg38 0 4547.1634476184845\n",
      "CTCF.hg38 100 4575.6657791137695\n",
      "CTCF.hg38 200 4603.5126516819\n",
      "CTCF.hg38 300 4632.83701133728\n",
      "CTCF.hg38 400 4659.5853633880615\n",
      "CTCF.hg38 processing sites: 7500 to 7999\n",
      "CTCF.hg38 0 4688.607177734375\n",
      "CTCF.hg38 100 4716.262281656265\n",
      "CTCF.hg38 200 4741.601805686951\n",
      "CTCF.hg38 300 4766.674676418304\n",
      "CTCF.hg38 400 4794.573518753052\n",
      "CTCF.hg38 processing sites: 8000 to 8499\n",
      "CTCF.hg38 0 4820.598433017731\n",
      "CTCF.hg38 100 4843.129068136215\n",
      "CTCF.hg38 200 4867.004532337189\n",
      "CTCF.hg38 300 4891.286687374115\n",
      "CTCF.hg38 400 4916.885512590408\n",
      "CTCF.hg38 processing sites: 8500 to 8999\n",
      "CTCF.hg38 0 4943.517446517944\n",
      "CTCF.hg38 100 4972.247307300568\n",
      "CTCF.hg38 200 5000.508421182632\n",
      "CTCF.hg38 300 5027.932875871658\n",
      "CTCF.hg38 400 5059.2655165195465\n",
      "CTCF.hg38 processing sites: 9000 to 9499\n",
      "CTCF.hg38 0 5090.296101808548\n",
      "CTCF.hg38 100 5119.797508478165\n",
      "CTCF.hg38 200 5151.841136932373\n",
      "CTCF.hg38 300 5176.662251234055\n",
      "CTCF.hg38 400 5206.314085006714\n",
      "CTCF.hg38 processing sites: 9500 to 9999\n",
      "CTCF.hg38 0 5235.855199337006\n",
      "CTCF.hg38 100 5262.063745975494\n",
      "CTCF.hg38 200 5293.228841781616\n",
      "CTCF.hg38 300 5329.128045797348\n",
      "CTCF.hg38 400 5356.906836748123\n",
      "CTCF.hg38 flipping reverse sites\n",
      "CTCF.hg38 taking means\n",
      "CTCF.hg38 mean of 10000 sites\n",
      "CTCF.hg38 smoothing\n",
      "CTCF.hg38 taking metadata means\n",
      "CTCF.hg38 exporting\n",
      "done CTCF.hg38\n",
      "debugging!\n",
      "analyzing: SPI1.hg38\n",
      "SPI1.hg38 total sites 10000\n",
      "SPI1.hg38 sites on selected chromosomes 10000\n",
      "SPI1.hg38 processing all 10000 sites\n",
      "SPI1.hg38 processing sites: 0 to 499\n",
      "SPI1.hg38 0 5386.627470970154\n",
      "SPI1.hg38 100 5409.657544136047\n",
      "SPI1.hg38 200 5433.393749952316\n",
      "SPI1.hg38 300 5458.665828943253\n",
      "SPI1.hg38 400 5484.5308537483215\n",
      "SPI1.hg38 processing sites: 500 to 999\n",
      "SPI1.hg38 0 5508.925076007843\n",
      "SPI1.hg38 100 5531.499399423599\n",
      "SPI1.hg38 200 5556.629217624664\n",
      "SPI1.hg38 300 5576.979330539703\n",
      "SPI1.hg38 400 5599.901841878891\n",
      "SPI1.hg38 processing sites: 1000 to 1499\n",
      "SPI1.hg38 0 5625.421909093857\n",
      "SPI1.hg38 100 5648.05535531044\n",
      "SPI1.hg38 200 5672.85666012764\n",
      "SPI1.hg38 300 5698.242681026459\n",
      "SPI1.hg38 400 5719.532389640808\n",
      "SPI1.hg38 processing sites: 1500 to 1999\n",
      "SPI1.hg38 0 5741.4881110191345\n",
      "SPI1.hg38 100 5762.460371255875\n",
      "SPI1.hg38 200 5783.663969039917\n",
      "SPI1.hg38 300 5807.438919782639\n",
      "SPI1.hg38 400 5829.568371295929\n",
      "SPI1.hg38 processing sites: 2000 to 2499\n",
      "SPI1.hg38 0 5855.116290569305\n",
      "SPI1.hg38 100 5878.617975473404\n",
      "SPI1.hg38 200 5900.8985431194305\n",
      "SPI1.hg38 300 5926.032073974609\n",
      "SPI1.hg38 400 5947.985265254974\n",
      "SPI1.hg38 processing sites: 2500 to 2999\n",
      "SPI1.hg38 0 5974.313254594803\n",
      "SPI1.hg38 100 5998.648306131363\n",
      "SPI1.hg38 200 6020.865496873856\n",
      "SPI1.hg38 300 6044.398082733154\n",
      "SPI1.hg38 400 6068.792575836182\n",
      "SPI1.hg38 processing sites: 3000 to 3499\n",
      "SPI1.hg38 0 6092.29595208168\n",
      "SPI1.hg38 100 6116.472281932831\n",
      "SPI1.hg38 200 6141.160546302795\n",
      "SPI1.hg38 300 6161.603942632675\n",
      "SPI1.hg38 400 6186.281160354614\n",
      "SPI1.hg38 processing sites: 3500 to 3999\n",
      "SPI1.hg38 0 6218.224892377853\n",
      "SPI1.hg38 processing sites: 4000 to 4499\n",
      "SPI1.hg38 0 6332.963740348816\n",
      "SPI1.hg38 100 6356.9019203186035\n",
      "SPI1.hg38 200 6380.175500154495\n",
      "SPI1.hg38 300 6404.48303437233\n",
      "SPI1.hg38 400 6429.55345416069\n",
      "SPI1.hg38 processing sites: 4500 to 4999\n",
      "SPI1.hg38 0 6456.112712144852\n",
      "SPI1.hg38 100 6481.960059642792\n",
      "SPI1.hg38 200 6504.489309549332\n",
      "SPI1.hg38 300 6530.693487882614\n",
      "SPI1.hg38 400 6559.685062885284\n",
      "SPI1.hg38 processing sites: 5000 to 5499\n",
      "SPI1.hg38 0 6587.384525537491\n",
      "SPI1.hg38 100 6611.969472646713\n",
      "SPI1.hg38 200 6636.325780630112\n",
      "SPI1.hg38 300 6658.9256503582\n",
      "SPI1.hg38 400 6680.394747495651\n",
      "SPI1.hg38 processing sites: 5500 to 5999\n",
      "SPI1.hg38 0 6709.115213394165\n",
      "SPI1.hg38 100 6731.966513633728\n",
      "SPI1.hg38 200 6756.301933526993\n",
      "SPI1.hg38 300 6779.019908666611\n",
      "SPI1.hg38 400 6803.710059165955\n",
      "SPI1.hg38 processing sites: 6000 to 6499\n",
      "SPI1.hg38 0 6829.321469068527\n",
      "SPI1.hg38 100 6851.334080696106\n",
      "SPI1.hg38 200 6875.145072937012\n",
      "SPI1.hg38 300 6900.301434993744\n",
      "SPI1.hg38 400 6926.254047393799\n",
      "SPI1.hg38 processing sites: 6500 to 6999\n",
      "SPI1.hg38 0 6953.959748506546\n",
      "SPI1.hg38 100 6979.536741495132\n",
      "SPI1.hg38 200 7007.721253156662\n",
      "SPI1.hg38 300 7033.439537525177\n",
      "SPI1.hg38 400 7057.8307185173035\n",
      "SPI1.hg38 processing sites: 7000 to 7499\n",
      "SPI1.hg38 0 7088.799530029297\n",
      "SPI1.hg38 100 7115.2164747715\n",
      "SPI1.hg38 200 7142.671052694321\n",
      "SPI1.hg38 300 7169.167088270187\n",
      "SPI1.hg38 400 7199.608825683594\n",
      "SPI1.hg38 processing sites: 7500 to 7999\n",
      "SPI1.hg38 0 7226.074929475784\n",
      "SPI1.hg38 100 7253.29844212532\n",
      "SPI1.hg38 200 7280.084933996201\n",
      "SPI1.hg38 300 7308.475376367569\n",
      "SPI1.hg38 400 7333.694406509399\n",
      "SPI1.hg38 processing sites: 8000 to 8499\n",
      "SPI1.hg38 0 7359.12029337883\n",
      "SPI1.hg38 100 7384.921423196793\n",
      "SPI1.hg38 200 7410.043523788452\n",
      "SPI1.hg38 300 7435.188150167465\n",
      "SPI1.hg38 400 7461.775425672531\n",
      "SPI1.hg38 processing sites: 8500 to 8999\n",
      "SPI1.hg38 0 7487.203869342804\n",
      "SPI1.hg38 100 7517.162256717682\n",
      "SPI1.hg38 200 7542.305043458939\n",
      "SPI1.hg38 300 7568.018602848053\n",
      "SPI1.hg38 400 7591.431878566742\n",
      "SPI1.hg38 processing sites: 9000 to 9499\n",
      "SPI1.hg38 0 7621.275614500046\n",
      "SPI1.hg38 100 7645.808431148529\n",
      "SPI1.hg38 200 7672.4874086380005\n",
      "SPI1.hg38 300 7699.558943271637\n",
      "SPI1.hg38 400 7724.826439380646\n",
      "SPI1.hg38 processing sites: 9500 to 9999\n",
      "SPI1.hg38 0 7750.075417280197\n",
      "SPI1.hg38 100 7775.079741477966\n",
      "SPI1.hg38 200 7797.862214565277\n",
      "SPI1.hg38 300 7827.343943357468\n",
      "SPI1.hg38 400 7855.695521116257\n",
      "SPI1.hg38 flipping reverse sites\n",
      "SPI1.hg38 taking means\n",
      "SPI1.hg38 mean of 10000 sites\n",
      "SPI1.hg38 smoothing\n",
      "SPI1.hg38 taking metadata means\n",
      "SPI1.hg38 exporting\n",
      "done SPI1.hg38\n",
      "debugging!\n"
     ]
    }
   ],
   "source": [
    "#cell for testing\n",
    "#start the timer\n",
    "start_time=time.time()\n",
    "\n",
    "#for testing use a single CPU\n",
    "for item in to_do_list:\n",
    "    final_data=run_full_analysis(item)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x7fb9d889d750>]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXUAAAD4CAYAAAATpHZ6AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Z1A+gAAAACXBIWXMAAAsTAAALEwEAmpwYAABUrklEQVR4nO29eZBl113n+f29fd/y5Z5Zm6QqLZa1lRWSZVsGG7A9NGJHptkaAodnWIahGdoEHR6GmRiGnqYnhsBt2k27aTOMjTE0CMJgjDfhtiSrJJW2kkpVlbVl5fby5dv35cwf9577bma+5S7nvtzOJ6JCWfe90jl537u/+7vf30aMMUgkEonkcODa6w1IJBKJRBzSqEskEskhQhp1iUQiOURIoy6RSCSHCGnUJRKJ5BDh2auF0+k0O3HixF4tL5FIJAeSF154YZMxNjno9T0z6idOnMC5c+f2anmJRCI5kBDR9WGvS/lFIpFIDhHSqEskEskhQhp1iUQiOURIoy6RSCSHCGnUJRKJ5BAhjbpEIpEcIqRRl0gkkkOENOoSiYO8sVrEX710a6+3ITlC7FnxkUQiim+8lcFmqYH7jyVwciIMl4v2eku4ka3i//jiG/j719cAAA8dT2IxFdrjXUmOAtKoSw48v/zZl1CotQAA9y8m8Fe/8Nge7wj4+FOv4dtXt/CBe2bw96+vYSVfk0ZdMhak/CI50BTrLRRqLfzMO0/gifvncP5mHvVWZ6+3hRvZKr7jzin82vecAQCsFet7vCPJUUEadcmB5lauBgA4eyKJx08rPY5W8rW93BIYY1gr1jETC2AmHgAArBYOp1G/uFbCNy9t7vU2JDqkUT8CtDtdtDrdvd6GI3CjPp8IYiGpyBs3c3tr1EuNNqrNDmZiAUT8HkQDHqzu8Y3GKf63v72AX/zsi5CzjvcP0qgfAf6Xp17Hhz/17F5vwxFuqcZyPhnEYioIAFjOVfdyS1hXvfJp1UufiwcPpafeaHfw/LUt5KstXM/u7TmX9Bhp1Ino00S0QUSvDXj9TiJ6hogaRPRr4rd4+Pj88zfxz//o2bF5N197cwMv3sih2myPZb1xcitfg8/jQjrsx1Q0AK+bcHNrb71irp/PxBSjPhMPHEpN/fyNPBpt5Qnw5eX83m5GomHEU/9jAB8Y8voWgF8G8G9FbOiwwxjDJ79xBf/tchYrY/DeVvI1rBTq6DIlZ/qwcStXw3wiCJeL4HYR5hPBPffUuVfOjfpsPICV/OEz6t+6kgUR4Pe4cP5mfq+3I1EZadQZY09DMdyDXt9gjD0PoCVyY4eVF2/kcHWzAgB4eQwXwrnrOe3nV5YLjq83bpbzilHnLCRDe66pc/llKuYHAMzGg9gsN9BsH664xjNLWbxtLo77FhLSqO8jxqqpE9FHiOgcEZ3LZDLjXHrf8IUXbiHodcPndo3FqL9wbQshnxvpiA+v3jp8Rv1WroaFZM+oL6aCWN7aW099rVhHMuRFwOsGoHjqALA+BglmKVPGn5+76fg6tWYH52/k8ehtE7hvMY7XV4pjuWk9cyWL55ayjq9zkBmrUWeMfYoxdpYxdnZycuCIvUNLvdXB3768gg/eO4O75mJj0SHPXc/h/sUE3r6QwGuHzKjXWx1slhu7PPVspbmn8YP1Yh3TqvQCQEtrHIeu/omvXcH//IVXkK82HV3nhes5NDtdPHpqAvctJtBsd3FxreTomgDw63/xMj72l686vs5BRma/jJEvvb6GUqONH35wAfcvxPHqcgGdrnPB0nKjjTdWizh7PIm3zcdxeaN8qIKlK7rMFw732pf3UIJZK9Y1Qw4Acwnl53Hkzz+rerFOS23PLG3C7SK842QK9y0kAADnHXZSbmSruLlVw9XNCq5nK46udZCRRn2M/OWLtzCfCOIR1bupNDu4kik7tt75G3l0GfDQiRTunY+PLVjaGlNevJbOuMNTB/Y2rXGt0NCCpAAwEw+qx5311G9uVbVz4rS098yVLN6+EEfE78FCMoh0xOf4mt+83CtyevqtoynfGsFISuNnATwD4AwRLRPRzxHRR4noo+rrM0S0DOBXAfxr9T0xZ7ctluVcFc9ccV6ne2U5j8fPTMLlIty3mAAARwNML1zPgQh44FgC987H1T04L8H8/GfO4Vc//7Lj62iFRzs0dQB7ltbY6nSRrTS2yS8RvwdRv8fxXPXnrir5DCGfGy87+DkzxvDaLeUJEACICPctJMZg1DOYiQVwLBXC1y9Koz6IkQ29GGMfHvH6GoAFYTsaI2+tl/B7/3ARX76wji4DvvmvvkPz9ERTb3WQq7Y0r/LkRBhRvwcv38zjR88uOrLmuetbODMdRSzgRdTvQTridzxY2mh38K0rWfjdLnS6DG4HOybeytfgdtE2r3gy4off49ozT32j1ABjveAoZzYRwGrB2RvNs0tZJENevPfMFP7p0iYYYyASf/6L9Taane62G9d9iwl89eIGSvUWogGv8DU7XYZvXcni/XdNI+Rz48/PLaPR7sDvcQtf66BzpOWXf/UXr+BbV7L4/gfmAQAv3cg7ttbO3GWXi/D2xbhjnjNjDOdv5PGQzpu6dz7meLD0gpoFUWq08eaas1LPrVwNM7EAPO7e15iIsJAM7pmnvqYa7ukdRn0mHnRcfnnuahYPn0zh/sUENssNx54MsuUGAGAi4tOO3TUbA2PAUsYZrfvCShH5agvvuj2Nx09Potbq4PmrudH/8AhyZI16u9PFhZUifuzsIn73h94Ov8fZFEPupek9uPsWEnhjtehIV8FSo41So40TE2Ht2L1jCJbqb4zPXx1Y3iCEnTnqnIVkCMv5vfHU1wqKwdM/PQDAbCzgqPyynFOCiDxeAzinq29VlMyaVNivHUurBp6/Jhqupz92exqP3jYBn9uFb7y14chaB50ja9SvblbQaHdx91wMXrcLb5uPO6pvcy9tVmeE3r6QQLvLcMGB4OVmSTEu6WjPm7pzNoYug1b85AQv3shhNh7AXDyA568560ndytW26emcxdQeeuo7WgRwZhMBZBwsQHpuSbmBPnJqAnfNRuF1k2O6elY13BPh3ndrQjXwWceMegZ3zkQxGfUj5PPg4ZMpfGMMwdI/efY6/ublFcfXEcmRNerckN49p8R071tI4NVbBceyNnbKLwBw26TiRTuRfrdZVi6udKTnTU1G/dtec4KXbuTx4LEk3nEyheevbTnW36bd6WKtWB/oqRdqLRTr4y9yXi/W4fO4kAht15Vn4wEwBmyUnPHWn7uaRSLkxZnpKPweN+6ejY3BU+8Z9ZTqqXNpRiT1VgfPX8vhsdvT2rHHbk/jrfWyo/n4lUYbv/03r+OXPvsSfueLbziafiySo2vUV4rwuV24bTICALj/WAINBwsoVgs1JEJeBH29wM5UVDHwGw4UpWyqF5feqPOfnbjwAOX3uJWv4YFjCbzjRAobpQZuOFTduV5qoNNl/T11nta4B976WkHpo74zQOl0WuNLN/I4ezypjfJ7u+qkOGGI+hn1sM8Nn8fliPyyWqij2e7inrleUt2JCeUzdrKnznNXs2h1GB49NYH/8PQS/uXnzzu2lkiOrlFfLeL0TAReNcj2gMMphvxi1xMLeuDzuJBxwMj2M+o8sLXpkFF/UdXTHziWxMMnUwCAbzukq6/1iVFwptWeK06c11Hw4Rg7mVP36VQTt7VifVvm1n2LCZQbbSw5UAeRLTcR9rm1NgiAEqCeCPsckV94e4X+VbrO3bj/6dIm/B4X/vO/eAd+8pHj+KvzKweieO9IGnXGGC6sFHH3bO/Ov5AMIhX2OWbUVwt1zO2QCogIkxE/MkUHjHqpARdt96aifuUmknVIfnnpRg5eN+GeuRhun4wgHvTi+WvOGPV8VZFWkiHfrtdiQUX6KO2R/LIz8wVw9qms3uqgVG9r8hoA3DUbBQBc3nDAqFcamNA5C5yJiM8RT71n1HtrzqpPPk4Gn795aRMPn0wh4HXjsdsnADhzPkVzJI36RqmBbKW5zagTEe5fdK7b3Fphe+k4ZzLqx0ZJvFHPlJtIhX3b8sSJCOmwzzEP9qUbedwzF0fA64bLRXjHiaRjwVI+aDoe3J0THQ0o5RfF2ni9KsaY+kS22+BFAx4QAcWa+BtNRv3+TPaR2rYc0Jy3Ks1tzgInFfY74qnz329KX48Q9cPtIsfkrLVCHZc2ynj3HYqOf8e0cpN8a10a9X3JhRUeJI1vO37/YgJXMmXhAbZ6q4NspYnZPo/lU1G/9qUVyWa5sU164aSjfkc89Vani1duKUFSzgPHkri6WUG5Id64DjPqMbX4ZdyB0mK9jUa7q3nlelwuQtTvQbEu/lzwm7TeU+eB2pwDRjZbbm7LfOFMhH2OxGvWi3UEvC5E/b1aSbeLMBX1O6ap8xTKd92uNB48ngrB53bh0rrzTcvscjSNupr5cqf6iMq5bzEBxoBXBaeC8cfHfp76VMzvSEbEQKMe8Tuiqa/ka6i3utvOKde7nbhpcaPOvXI9IZ8bbheNXX7hXng81L+iMh7yavsWCU9f1Rt1v8eNiN/jiOc82FN3Sn5R2i7sDj4HHNPUv3kpg3TEhztnlO+zx+3CqckwLkqjvj+5sFLEsVRI8+g4b1Oj66KbXnHdj+uAeiYjAeSqLeH5y4pR7+9NOWHUufGY7JNC6ZRRj/g926pJOUSEWMAzdvmFPxns/F5x4kFnjHqmT1AcAJJhr3BPnTGmGPU+361U2IdqsyO8mG6jVMd0n6ef2bgzBV2MMXzzchaP3Z7WsokA4MxMFJek/LI/ubC6PUjKSYV9CHrdWqc7UfQKj/p76oD4jJTNUnOo/CI6f5xLOnoPjq/vxE2kUGv1lV440YB3/PKLehOJBfu3VHLMqJd2l+0DQCrkw1ZV7HrlhtL3pZ/8wp0I0U8HG8UGJvvEKWZiSusF0d/lyxtlbJYb2/LiAeD0dBS38jVH5ESRHDmjXm22cS1bwV19jDoRYT4ZFN73ul/hEYd7tiKDpZVGG7VWB+lonwyFsA/tLhNuXLYqyv71Rt1JT71Ya2tZLv2IBT0oOaBfD2PPPPVSA6mwT0vP5STDPuGeer8WARx+bEtwzGa92N9Tn0sEUG12hMcpltSK67tmttuIO6aUmpb9rqsfOaN+K1cDY8CJdP9ujPOJoAOeeg2xgAdh/24PjnvqIg1fvxx1jlNVpVrpuM5bTIZ8cJEznnqx1kJ8gEcMKIbViUyTYRSHBG/5cUc09XJjm+zFcULj5t+bnU8FfD0A2KyI+7zLjTYqzc62dEaOlqsuWIK5qRbMLewobDutZsDsdwnmyBn1lSH6NqD05r4luGx/pVAfuJ6WvywwWNoz6v0ekZ2RRLbKTQS8LoR82zMUJhwKzI6WXzx7kv0CDPbUYw566voeP5xUyIec4JTGrT59Xzj8mEhPnef1T/Ux6jwQL7ql8XKuhojfs6vVw2IqBL/Hte+DpUfOqK/mB1ciAoqnnqu2hFaOrRXqffV0QPF4iBTdUBSZ0u6+L/r1AAhPa9yqNLWmTnrSEWdSNkcZ9VjAO375RTXYkT4ZOYDiqTfbXeGBxMwATz3pQOCyn8zGSTnQqXFdvS76yS9OtV5YzlWxkAzuyrZxuwh3TEfwljTq+4uVQh1E/dMLgd5oNJG6+mqhPvAm4nW7kAqJLQja7JO3zHHKU89Wmn0fydMRHzIO5MWPNOrBPZBf6i1E/Z6Bg0H4fkV664wxZEqNvp81N7wivfVeh8Y+BVZ+D7xuEhoo5U+wUwNqPIjEV5Xe3KoNHJZzemr/Z8AcOaO+mq9hMuLfFVTizAseXNxsd7FZbmAm1l9+AdSqUoGeOjfY/bwpp3TuQbnLk1G/lkctima7i1qrM1J+qTQ7aI9hVipnZPA2IN6oV5od1Fvdvkadt1AQ6TlvlZsIet3bGtNxlP4vfs2bFwG/LvrJL163C5MRv1D5hTGG5VxVG4u4kzumo1gr1h2R0URx9Ix6ob6tp/lOuKcuKljKC48GeeqAYvgygjX1ZMjb98bldhFSYZ/wQOlAox7xI1NuCE07G1ZNyuEGdJzpZ8V6q28xFMcJT51LW/2kNv55CDXqAz5n/Zoipb31Yh1Br3tbNake0bnquWoLlWZH6/S5k9PTSgbM5Y39K8EcOaO+UqhpHfP6MaX2lBAVLOWpiv3ybHtrBsRmvwzIUedMhMUGLxljSpOnAZ56s90VmnbGjeLwlEa1VcAYC5CKtdbAICmgM+oCc8czfapJOamwsp5Ioz5IZuNMRMR2alwvNTAd8w+ctToreEzgoMwXznF1ktheDWExwpEy6rzh0qBMFEApB56JBYRp6sOyBTiTUbHe7KAWAZx0VGyPjqoqAfTLXXZCwzdk1HlTrzFmwBTr7YGFR4Cznvow+UVkrvqgmzdHdBrlRrHeV0/nzMQDQo06l10XU/099Tk14UF02rNIjpRRL9baqDY72gcziPmkuFz1YdkCnKmoH60OQ06QB7dZbvQtPOIo/V/EPpID/W9cWl68wCcRbqhHVZTq3zsODHvqAo26FhTvcxNPhJTMKpFVpVvlZt+bN0e4US81MDXkuzwbDyjzeAV9zjdzwz31kM+DVNgnjfp+YUUbrDDYUweAhYS4XPVh2QIc0QVIm+Vm3xx1zkTYL9RTz2pVhoPz4kVm94wq8gF6pfpjlV/qLUOSkGhP3e2ivn3l3S5CIiiu/4sisw2XX9IRP8qNtrA0yvVifdtwjJ3wLLZ1QX3ql3NVJEJezSnox1xC3JO8Exwpo86j5INyxjnzySDWinUhmRO5ilKU0y9bgNNrFWD/i1lrdlButEfKL5VmB7WmmAuPP430u9id8NTNBErH5al3uwzlRluTffrhVtvvijbqE2HftsZTepJhn7Ce6tVmB412d2SgFBCj45cbypN1v2pSjuhhGTe3agODpJz5hPhWIiI5Ukad916eG+GpzyWC6LLeZHg7ZAcU5ejhmqGItMZhj+OcdFiszs2zHfr9nomgF24XCfXUeaDRiFEfVwFSqdEGY8N1fkB8/nym3D9HnZMKiev/0m826a71BBp17n3360/P6VWVCjLqauHRMObUJ3mnhqrb5UgZ9dVCDR4XDb0IAF1aowAJZlQKGABNMxRh+LQ2rH3Kxjn8NVFGXbvY+3jqLhchHfFhsyROZy3UWgj53ANrDYBeVee4CpD4OsM0dUB8/5dBhUecpECNO2sg6M9fE5EBsz6kRQBHZCCeMYZbudrAIClnPhFEpdkZe2tnoxwto55X9LlBFX8cXoAkIhiSM2DUw34PQj63EE99mNfM6V0I4jw4n8eF8ACJSfRgjlHVpEBP6hiX/MKfCIZlvwDijfqoTCeR/V+4x5805KkLcFBUyW6Yph70uRH2uYU4DZlSA412F4sjPHXRtSyiOVJGfaVQG1oExBHZKiBrwKgDfGiv/QuBG4x+gbPeWooREBUsVSQm38BcYp6yKYrCiCwTTjQwvva7o9ruckQa9W6XKR0aDXjqIqQCI1lHve+WGCML9OTCYWtmBVw7N9Un80EtAjhz0qjvH0ZVk3ICXjcmBKUtGZFfAHEXuyYDDPEY+SOySPll2O+YjohtFWDEUwfG2/+laCB3HhBr1Au1FlodNjR+kgp70eowIZW1RiSmWEDpfSNC8slXW3C7aOTTTzoiZprXsprOOKhFAGfOgf5QIjkyRp0xNrSx1k6UXHV7wZd6q4NqszNWo96b3Tn4wgt43Qh4XcKqPEc9jfC8eFGBpUJteOogJzbG6Uej2u5yRM4p5YZsWIohzynPVQQ4DOrvOKwVAh8lKOIJaavaRCLoHfgEyElHxAxTXzboqacjPvg8Lump7zXZShPNdtewUZ+K2m8ZayRbgCPMUx/RKZATDXiFFWxky8OrDCejfjQ7XWGBpVK9bchTH6v8YuAJCVA82Yag9rv8hpUYIrVprQIE6OrFegs+twsB7+D0XEB5WhHx3cpXm0P1e46onv1rhToSIe/I34+IHBmmI4ojY9RX88OHY+xkcuxG3YeCAKNn3IsVN5hZkV+GBWaV3z9TFpN2Zkp+GZunrvZSH9B4isP3LUIW0maiDvGcRbYKKI1og8BRBpTY/27lKi0kQ6M/58mIEjfodO09CWZGVK/q2c+56kfGqPNq0lEtAjjpiNJC1M4XxUjfF47iqduXKEa1f+WIGszMJaZhEkBvVql9w9LudFFuGPPURd64RlGstRHxe+AZkmYJ6BqNCTj3WnB2yLkQmTdeqreHynocUaMEc9Xm0KcQzkTEjy6z3zc+MyKTSM9cIiB8QpoojoxR7008Mu6pd5m9i2HLQAoYJx5UAlo1m4/lo2Z3chQv1r7BM5K7PCmwVQDfs9HfsVRvjaVIpFhvDfWYOSL7vxgJXPLvnoi0RqW3jYHzLshhyFWbhjz1tKCMm1GZRHrmEkFslBpotMVOsRLBSKNORJ8mog0iem3A60REv09El4noFSJ6UPw27bNaqMPndhnymgGdIbIhwRgxeBxRF7tiXAzqzQIMC59HOUxi4t6WyEBw3MDFHg140GXKIAmnKRqUvYQadQOBy6jfA4+gbJRSvWXIUxcRy2BMaXA3LDWXw58S7erqmVL/sYD94GnP6wXxoxrtYsRT/2MAHxjy+gcB3KH++QiAT9rflnhWCnXMxAMDe2TshHc5tPNF2aoozZaMGFlRF7u5zBARnvroDIxecy1xRt3IOdX6v4whrdHozVS0p+73DA9cEpGwqtJivT30BsIRkUpaa3XQbHcNPeWKqCqtqH1mhnU31cON+nK+anlNpxhp1BljTwPYGvKWJwB8hik8CyBBRLOiNiiK1byxwiOOCE99Sw30GLmRiBqgUDQcRBRTbdkLBg++GPweNYVSpKduMFAKjKf/ixLLMCG/CGiHO6orpH5NEZ91yeCNKxbw2h4lyNtQG5NfuKdu/cZlpGeSHl51vmIz7dkJRGjq8wBu6v6+rB7bBRF9hIjOEdG5TCYjYGnjrBbqWtGAEdIC+rFsVRqGMl8AIBGy78G1Ol1Umh3DF56IyfaaUR/xmCxKZzVj1KNjHJRh1FPvtd8VUQw0vCskR1RqZ7Fm1FNX3mNnTZ6tYyRQGg964XGRLU992LCRfvCWv/sxA2asgVLG2KcYY2cZY2cnJyfHtm6ny7BWNF54BABhnxtBr9tWJaTRalKgZ6TyNox6yUwQMWD/wtP/+8iIi115JBeTsgkY9NTHKb8YlL28bqVHjqiaBOOZTvbOfaujDPs2uh5g06irgV0jmjoRKWP0BBh1o9kvfo8bk1H/vsyAEWHUbwFY1P19QT22b8iUlNREIy0COERku2eJGaPem6lp/WI3MuaNI2oyUKmupPKNKnaKBcT0ETdajq9/j9PyS7fLUBrRS12PuEIzYymGiqdub72ygaAsR8QoQS6/8OKpUdid5qXJLwY9dUBp+yuiPbdoRBj1pwD8lJoF8wiAAmNsVcD/VxhajroJTx2w31PCjFGP+j0gsie/GJkIxBHxiKz8+5ahC12UrmskOMgZl/xSbhrrpc6JCTLqJcMphvbz9fk5NJSnLsBByVeNyy8AbxVgz1N3kbFCQc5MTOx8VFEYSWn8LIBnAJwhomUi+jki+igRfVR9yxcBLAG4DOA/AvgfHNutRcxWk3LsVJV2ugz5WmtoAFGPS82SsXOxm/HURUkTpTFmRADKMAojhgXQGXWH5RejvdQ5os6FGfnFrqeutRY2mKeu7M/6jYTHahIGb5QTEZ8tTz1TbiAV9o984tQzGw9o09T2EyM/IcbYh0e8zgD8grAdOcCqyWpSTjrix7evDkv8GUy+2gRjQMpA9J6TsNnsyUhrVI4I3RMASg1juct2b1icssGbCKDLunFYftHK9Q3EMgDFMNrNmmCMqYFSA5+1X+k302x34fNYezjnNyGjcg9g7wkpX1WeAEdV6HIm1f4vjLGRDcD6MWrYSD+m4wEU621Um22EfMY++3FwJCpKV/J1BL1uQ8ZOz2TUj1y1hZaF1KzeNCDjX5R40Iu8jVQ3UzncQTHShFFPPa5WsNqt7iw32iP7q+iJ+L2Oa+pGe6lzIn6P7Va4jXYXzU7XcC8WALa8da0LpcFKXsDeE1Kualy6BBRPvdHuWj6vmXLTtFHniRf7TYI5EkZ9tVDDbCJg+g7OP2Qr5cdmqkk5dgNo3GM046mLkV+M3UQ6XYaqzerOcr2NsH+0ns6JBuwb0FGYCd4CSqaQ3T2ZkXxEPJWVTNy4eHzIzhNSrtoyrKcD9lsFbJYaWr67UWZiipwrjfoesFKojxw23Q87lWra6C8TX0y7Wmuh1oLXTQh4R3+sYZ8bLhpfoDQmKtum0UbEb/yJK+L3oOK0UTeRGQIoTw9lm+fdSDMvTlRA+qqZ39HlIkR89jJuchVjfV84EzauVcaYJflF9NBrURwJo262mpTT6y5o/ouieeom7v62PfW6Uk1q5ImEiIS0pjVTOg7YL48vN4zdRDhhv9u2AR1F2WDbXU404EGz07XVDKpoInDZ89Tt1ECY+x3t1iUozbzMeOrWq0qL9Taana7halIOL0Dab2mNh96oN9tdZMoNUznqHDutArYseOoJ1ahb1Z2Nzu7k2K00bLSV/hxmep7YTa0r181r6k7LL7xhWNjgvvj+7dxszEg+vcClnfXaCPvchgOXSk91e4FSc0bduqdutpqUE/C6kQh5pfwybtaLdTBmPkcd6H1RrBQgbVWaiPo9prIN4kEv2jZ0Z6NVjRy7fa9LpgpSxGj4lUZnZPWqnojf7bhRLzfa8LgIfoOftWbUbezLjKceD4rx1E19t2xMP2qqAU8z8gsPqlrR1DWjbtJTB5RcdSm/jBl+wq146kGfG1G/x7KnnjIZeLHbKsCsUbfrTfXS3Iz3A7EjvzTaHTQ7XXOe+hgCpZVGG2G/x3AgPiJC4zYVKBWhqZuTvewUPGmFRyaSDLxuFxIhryVP3Uo1KUepKt1fuepHwKhbqyblpKPW5h+aqSbl2O3gVzQ4u5MTC9hL99M8dQOBSxGBUjOl6pywgPTBUZhNs4wK8dSNyy98b3Y/a6NFX4C9Bm5mOjTqSVucVWpVfgEUXV3KL2OGF3nMWDTqkxFrVaXZSnNk58Kd2O21XTBYNs6xm21jRn7pVXdaNyzcCJo1oE218MYpFE/deJol99Ttaept+NwuQ5KPx+1CyOe2Kb8Y720D2Ptu8WZeZq+fibBPS1AwQ6bcgNdNputYACWtcbPc3FcTkA69UV8t1BD1e0x5GXrSUZ8lTT1nwVO3kyGiVBga66XOsRsoLZnoB+JxuxDx25N7tI6QpgKlynudTGusNDqGg6SAKE29hVjQuORj97MuGpx6xImpslfXwoxfM2139aQsDgPJlJTZpFYqUXlW3UZx/0xAOvRGfSVfx6zJ9gB6JiN+0+13GWOWNHXeU92Kh1NtdtDuMtOB0lKjbXm4thlPXVnPXqdGzVM3Kb/o/60TmK5y5Rq3HaNuOtPJi1LDpqdusA0CX08ZJWj+d9TkF4MdGjlWjbqZ2aQ7sZLW+K7f/Sr+4KuXLK1nhCNg1Gva6CkrTEb9KNbbpoZJlBtK3us45RczfV843BhbNXhmy+Ptyj1lExo+x+7vaIRKo42wid4ffP925JdSvY3omJ7K+FOgKU9da0NhxaibTwcGFPklV22adlK4p24FswVI+WoTy7ka/B7jcp1ZDr1RX85VsZAMWf73WvmxCQ8gV+G9oM19KXlf8nzNvLdhpu8Lx26PDqMDMvTr2QqUqobZjH49Dk+dZ78YJeB1we0ilG14zsqkJXOes9U89XqrqzwFmgyUAta+W/lqE0Gv21B7ZT2psA+MmXeKzAyc3sm01v/FWAbM9awy0/T4hHWbNIpDbdQLtRaK9TYWkvY8dcBcAZKRYcz9ICLLEoWZvi8cuxkppbpSkGK0XanSqdGGd2pBfhFR6DMKRX4xboCISG1fYKOi1EL6aslqqqwWOzF3EwGsZdzw2b5m4c3ztirGr9VulyFbMd/MixP1exD2uQ176teyFQDAiXTY0npGONRGnY+asuOpWzHqVqpJOUqrAPMXQq+XurlcYsB6qlvJbPAs6Dl08gtjDJWmuUApoNxs7PZiMZWNEvBY9tRLFoy6Jr9Y9NTNBkmBXraMmQIkLteYbebFISLMxANYN6ipc0/9WEp66hpfeWMd7/ydr+CWgYGv/D3zNjx1K+XHW1qHRvN3/3jIZ9FTN6+pi5BfzFzodqcflRstuF3GGpZxnJZfGu0uOl1myajbkl+sBEotnvte293xPAXmay0tacAMXO40EyzNaIVH1pMpZuNBw5769WwVM7GAaWnJDAfOqId8HqwU6riaqYx873JOuSvakV+4hGLFUzeb/QJYb+plRVO3W2lYMtlcixc7Wc22qTQ6iJio3AScT2m0kjsP2Kt0rbc6aLS75uQX3aAMs/SqV83lqQPWvlsFk6m5HH6tmol/bZaU91qVXwAlA2bV4NCT69mKo3o6cACN+m2Tiha1tFke+d7lXA0Br8tUT/Od+D3KcA2znrpPnRpvlnjQi0LVSqc5C4/IAjR1c/KLvawPPuTaDDwrxalBGRUteGvBU7dxHgBzRtbOoIzeeuYdBitPgQWLnjqXO3OmPHXFGFuVXwBFSlkr1g1lyF3LVnFiwjk9HTiARn0y6kfY58aSQU99IRmyVFSwc02znnoq7LO0bjxoLVBaqLUQ8Rsf/wXYr/K0Ir8A1m8iZtvuAkpv77DPPQZP3dwNPBLwWM5TN9MigGMncGlm6DTH63Yh6HWb/qwZY0pltAVP3edxIRrwmPLU7bQI4JxUg55XN4fbpHKjjc1yA8fT0lPfBhHh1GQESyNOIKBo6nakF47ZVgFW+r5wkqqmbrYSr1gz1/cFsF8+bjpQGrDX1MtskQ/Hyf4v3Ns2M7gDUOQQq5662UHXgD2pzWyRGUcJjJtbr95SJCIr8gtgvgBps9xEwOuy9L3iGDXq13nmi/TUd3MyHcZSxpj8YqfwiGO2qVfWhlGPB5VKPLNeXKFm3osF7DVeMp2BYTMwW663TaUzcux4xaPgFZNmcucBe3NKNTnETKaTjfa7pboSoA6ZlBNjFqpYCxYC/nrMGnU7LQI43KiPskk3xpD5AhxQo35qMoxb+dpQDatUbyFfbdlKZ+SM01PnqVx5k7o6n3pkFquVhnxAxlg1fIueetTBkXZlNdfcSqC02uxYChqbreQF7A3KKNYUmc2s4YtaaL9r16ibbeplZYzdTsJ+D2ZigZHqwbUxFB4BB9aoR8BYL+ezHzydUYT8ko76UGl2UDXYx8JKMy8OL7rIm2y/a7YYhWO1yrP3SG58zbjW28aagS2b1PA5YRtSxyjsBEoBa6mW/PxZSTG06qmbuYFoa1r4bonx1I07YHaqSfWcmgyPjPNdz1aQjvgsNxc0ysE06gYed3qFR2I0daCX/jSMRruDUqNtw1O3NijDbIdGjhVvCrAoAQjQ1M30WOHYkTpGYdWo2ymKsuOpWwuUWruZWpmsxb8biaC16ycV9iNXMT4ScrPcQNqmpw70JOFh617LVnDcYT0dOKBGXdOwhjzuLOfsFx5xtKrS8uhcVKt9XzjxoDX5xex8Uk7MYlGKVmVoIkAY9nngImvyS0cd82dJU3cyUMqNukm9OWKjqRfXuM0UYdkZlFEyOfWIY6Xmgn/v7cgvzU7X0Ofd6nSxVW0K8tQjKNbbWofJflzPVnHcYT0dOKBGPez3YDrmH/q4s5yrwu9xCfnAtFmlBjz1XjXp+OSXdqeLSrNj6UKIBa2Vj1vJiHC5SGksZcFTt1rkAzg70q7SaCPgdZlKJeV7AmCpqpSnkprRuO1kOhVrbYvyi/LdMjNI3a78kjRRVbpVaYIxe+mMnFHqQb3VwWqhLj31YZxKR4YWIC3naphPBm3nqAPAlOapj9bqtL4vNrJfgF77USMULUghHF4+bubCA8wNyNCjtAowb2C5UbbiMUbUQKnZ39EIZbXK1cqeAGuec9lCERZgPShuNnWVEw960ekyUzfUYq0FImufM9BzpowES3nyg9W2u3pGqQc3t5T43wmHc9SBg2zUJ8ND80KVHHUxJ1ApJIKhYRlah0aLRt3jVgoozHjqVvq+cGIBL1odhnrLXPl40UbushVNvaJ56uZ/x7Dfg1aHoeHASDuzbXc59jR1c5W8vTWtDcowOyCDY2U+QKHWQtTvgctg58+daP1fDDT1ytgYOL2ThWQQXjcNVA+uqB680znqwAE26ifTYeSrrYGPWcs5MYVHgGJoUyFjY+14ibJVTR1QgqVmLwTAXOCM00t1M3exWykdBxQN3k6puhVNnf+OTqQ1mh2QwbHTErjcaGnDq81gJSje6TKUGtZuIlaNupUOjRzNqBt40uWe+pQAo+5xu3AsFcLVAerB+ZsFeN2EMzNR22uN4sAa9dsmIwD6a1j5ahNblSYWBXnqgPFWAVuVJojMz1fUkwj6TMovqqduoV+G1aIU/n6zRtZKlSFgT1PnRtcJXd1qlWvEhqdutj0Dx0qnRr4/M0VmnF6xmfHf0WozLw5v6mVEU+cFhSLkF0AJlg7y1M/fzOGu2Zij3Rk5B9aoD9Ownru6BQB46HhS2HrpiLGq0myliUTQa3hwRD8SIa8p+cWOp95LMzRnXMwOyOBYbQFbtij3APYM6CgqzbbpalLAXqOxcsNaZa0VTd1KSwKOFU89b9OoB71u+D0uQ0Y9U2og4vcgaKHxXj9OpcO4nq3uKijrdBleXS7g/sWEkHVGcWCN+kIyiKDXjVeXC7tee+ZKFgGvS+hJNOqp56rWC484iZDPVEqjlalHnKjFohSrwTPeftcsPEvETlDSiQKkSsP8gAwAcKuNxsbpqVsZlGG174uynvm2EHY9dSJSqkqNaOoCqkn1nJoMo9npYmXHrIfLG2VUmh1p1Efhcbvw6G0TePpSZtdrz1zJ4h0nUvB5xP166YgPm+XGyAyKbLlpaTiGnmTIa6r4yMrUI07c4oBg6xKA0ofFbHm8HU3dTvXmKKzKL4Caamk5+8VioNSizGalWpnLgWY8dauV0XpSEWNVpaKqSTmnpxW9/OXl/Lbj52/mAEAadSM8fnoS17NVXNNJMJlSAxfXS3j0tgmha01G/ai3Rhc1bFWaSIbtfSkTatGG0U6NxXoLXjchaEGvi1rwpgB7Rh0wb2B7RT77TH6xmP0CWCuKarQ7aHbM9dzhWBmUYTXLCQAiarGZUaPO2+7a8dQBparUqKaejtp7qtZz73wc8aAXX7+43dE8fzOPWMAzlswXwKBRJ6IPENFFIrpMRB/r8/pxIvoKEb1CRF8nogXxW93N46cnAWCbt/7sUhYA8M7b0kLXMjqrdK1Yx3TM+mgsQBlpx5jxjBReTWolJ7/XE8Ssp25RfrHYqbFcbyNkQcMHnPPUu2qVq2WjHvCa7h5pRw6xMijDaj0CoBSbxUxUldZaHbQ6zNKADD0TYZ/h7BeRnrrH7cK770jjG29ltjlkL93I477FhOU0TbOMNOpE5AbwCQAfBHA3gA8T0d073vZvAXyGMfZ2AL8N4HdEb7QfJ9JhHEuF8PRbPaP+zFIWUb8Hb5uLCV2rN6t08Jel2myjVG/bNupmq0qt9n0BgIDXBa+bTKc0Wu8HYi1AaEvmcGikHW+7a3ZABkfpqW7+5gZYNermb+BWRtnpMTOXln/f7XrqyZBvZJ56vdVBsd4WqqkDwHecmUKm1MCF1SIAxSa8tV7CA2OSXgBjnvrDAC4zxpYYY00AnwPwxI733A3gq+rPX+vzumM8fnoS37qS1R4pn7mSxcMnU6bLtkdhxFNfU4fPzsbtGXXuqRhNayzUWohavBCIyLLWajVQCljIi7eY8QEAIZ8bROIDpRW17e445RcttmBBU7cyN9RKN049Zvq/2G0RwJmIKF1Va83Brbl5xamodEbOe1T14OsXNwAAry4X0GXA/ccSQtcZhhHLNw/gpu7vy+oxPS8D+EH15x8AECWiXaI2EX2EiM4R0blMZneA0wrvOT2JarODc9e3sFqo4epmRbieDgDT6rTxteLgpl78tRmbnrrWU93gxVCsm596pCdmoSjF7IAMjtWxauV621LBDaDcuCI+8YMy7OTOA4rWz28MRinZyAKyIr8U6y0EvC7LSQd7YdS5U7VSqA18z5r6mt2n6p1MRv24dz6Or6m6+vmbeQDAfQsJoesMQ5Q7+2sAHieilwA8DuAWgF3fVsbYpxhjZxljZycnJ4Us/OhtE/C6Cb//lUv4iT96DgDwrjvE6umA4j37PS6s5gd/UdZVoz5t11Pn+b0m5Berj8eA+awIKwMyODxDx6ymXrHhqQOKNy1afrFt1P0eC09I9jV1M5lOZoeL7yQWGL9R59POdqYW6hHZxXUn33FmEi/dyOHvX1vFH3ztMu6ciWJC8BPBMIwY9VsAFnV/X1CPaTDGVhhjP8gYewDAb6rH8qI2OYyI34OHT6bw7NIWQj4PPvHjD+LOGbF6OqB4e3OJIFaHeOqrBbGeulH5xY6mDpjv1GjnkdxqXrwdTR1wplOj1V7qnKi6JzONxuxo6lYGZZQsPpFpawaNd+UUZtRVQ81nKvSDD9ERMe5yJ4+fmUKXAR/9f1/ETCyAP/rps8LXGIaRT+t5AHcQ0UkoxvxJAD+ufwMRpQFsMca6AH4DwKdFb3QYv/cj92Oz3MA9czEhXRkHMRsPDPfUC3VE/R7LFzmHf6mNBErtTF/nxAJebBRHz3zljNtb5Gta0ZE5in5tTuoYhQhPvctgKoPGTjaKlUEZRYuxEw6XXxhjI69NrTGdzeyX6VgALhrtqSdDXtvXaj/uX0zgWCqEhWQQn/yJh2zfpMwy8jdijLWJ6BcBfAmAG8CnGWOvE9FvAzjHGHsKwHsB/A4RMQBPA/gFB/e8i5l4ADM2JQ8jzMaD+NaVzYGvrxXrQvbhdhFiAY+hqtJaq4N2l9n64pgtH7djWLxuF4Je8329i/WWpeIqTsRCpsko7Hrq+vx5o/8PW33lLbT7tRo74cSDShfQWquD0Igag3y1BRcp+e128LpdmIkFsDzEqN/KieviuhO3i/DlX30PfG6Xo07mIAydPcbYFwF8ccexj+t+/gKAL4jd2v5jLhHAerGOdqfbN7tmrdgQdnNJhn2GAqV2+r5wYgFzsyTteOqA+aZeXbUntx2PMeL3mBoeboSeUbeY0qiTQ4wG7EqNNvwea4FLK4MySrWWrW6nca0uoT3SqPMnThH53HOJ4AhPvYo7ppzrmOj3ON+4axAHuqJ03MzEA+gyYGOAcVgv2C884iSCxpp62en7wokGvKg2O2h3jFUa9jx1q1qyub7epUYbjFnPlQbU4dPCA6WKnGNVfrEeuLQTFDf3VCbCUweMVZWKqCblzCeDmm6+E8YYbuVrjgRJ9wPSqJtgLq58CXhAVE+708VGqW47SMqJG2zqZafvC4f/W6MXuzZpyaLnbFXusRM3UNYUL7+4CJbaMwDWCrHKNrNRoiafyop1a7NvOfy7NW6jPpcIYq1Q79tjKFtpot7qCpu3sN+QRt0EswnFYK/2yX/dLDfRZRAnvxhs6mVn6hHHbEGQbfnF5JxS/jRix2OM+D2oNDtCR9qV1QEZVnVTK5lApXrLVhaQmRuqndRVzp556okgWh3WV3LjWTFOZL7sB6RRN8Es99Tzuz11UYVHnETQq01RGoYITd1sVoQ2IMOG7GA2AwOw9zuG/R50uubH9g3DTjMvwFo2it3UTjM1CXarSQFzRt1uaq4ebrD7STA8R92pQOleI426CWIBD0I+d99KNd4iQJSnngj5UKyPblGrTT2yladurslWSW2uZbUVQ8xEPxD9vuzIL050arQ6IINjzVMfn6ZeFCDtmTHqdgdk6NFy1fsY9Vv56rb3HDakUTcBEam56rs9da2aVJSnHjJmaPnFYvdCB4wH7JS+L/bWs1LsZMdT5023RBr1cqNjy2sO+9xwkfleLHYqa80MytA8dRv1AUZbO4tqu8uZG1JVupyrIRrwjD1/fFxIo26SQVWlq4U6vG5l6ooIjDb1KtaUx3E7DcysaOp2S8eb7S7qLWPFQEWb2TZArwGWyFYBVpuacYhIbRVgLmhsT2ozLr8UBQSo3S5CNOAZ6alXmh10bNZb6In4FaPdr6r0Vq52aPV0QBp10wyqKl0v1jEVDQjrmdxrFTDaU7cTQASsyS92DKzZrA8eKLWzJpdJrIzSG0TZ5nkAzGWjMMbsa+omBmXYDYhzjATGN0tih0ADiq7ez1O/la8d2swXQBp108zEg8iUG7suirWCmGpSTipkbCq6Umlpz7sxW2lo10PVbiKGnwxaCNvQ8IGehCBSflFaF9g16sY99Wqzgy4TI7UZ8dbt1iNwjHRqzKhD3dMC+5vPJXbnqjPGsOxgNel+QBp1k8zFA2Csp6Fz1ovictSB3pd7szy8ClLETEe3ixD1e0zKL+PL+rDbfwToBUpFyi92q1wBPojb2HnXWgTYbLAFGDv3Wiqpze+XEaPOPXWRk4gW+hQgFWttlBttKb9IeswmdhcgMcaEjLHTk44onvrmiNJ2UcElU1kR9bbttgSAcbmnWGvbysAAdPKLIKPeUVsX2DGwgLnzLiLF0Ew/+1K9BRLQi8Wcpy5uZuhcIoBSvb3NWbmZUzJfpPwi0ZiL7y5AKtbbqDY7tice6fF73IgFPCM99ZJNA8sx0yJVCdbZy5VW/j/GPXW7v2NUcKCUj7KzG88wZ9RVOURIbryBvHFVXrIbJzLqqbsImAiL1NQViUUfLNVa7kqjLuH089SvblYAiP+ipKN+zYMZhNIEyX77UCXNcPSF3mx30bBZZdhLoRyP3AMos1hdAkfa9cbK2Q+UmpVfxpW+KuJmCiitdEd91plyA6mwz9Jg8UHMqRXgeqN+2AuPAGnUTRPxexD1e7ZlwHz1zQ24CHj0lNgxeumIH5ulwYHSWrODcqMtJGNA0XaNPZID9iSAnq5rPLXOrq7L0wdFBUrLAqQQ5d8rnrqR9gXajcRW5pHxc1+s2b+ZKmt6UG910WgPTmHNlJrC54XePhWB20V46WZOO/bSjRzSEb823P0wIo26BeaTQVza6A2V+Mob63joeBJJQTnqnMmIf6j8wl8TMRHdaJWniDQ3s0U3yrg++xehSKOutUoQkNLYNti+oCzg6cBMkNpuTjzHSFVpptwQ8j3WEw148dCxJL6uzgttd7r4p0ubePz05J70OR8X0qhb4LvvmcEzS1ncyFaxWqjh9ZUivvPOaeHrTI6QX3gLYBEXg1FtV0SwjoiU/GwDGj5jTIj8Aqgj7UTJLwKkEP2/N6Zx239KMpO+Kuq8x/kg9SE1F5ulhtDMF87jZybx+koRG6U6Xl4uoFBr4b1nxMxH3q9Io26BH3/4GFxE+NPnruMf39gAALz/rinh66QjPpTq7YGVlxmBaWC8QGSUDCAqd9noTYRPdrIrvwC8U6NYTd1O0BIwp3HbHZ8HmBuUIUL2AoAp1ekYNKSEMYZMuSE0R53DDfjTb23iGxcVmfTdDgym30+IH9B3BJiJB/A990zjz87dxF0zMRxLhXD7VET4Olxj3Cw3+gZ2uBc/JchT7zKlXHuY0SgKkF8A49OWem137RuXsMmS/GGI0tTNaNylehthn9t2MNFoUFyUp86/nztrO7R1Gm00211HPPW7Z2OYjPrx9YsbuLFVxQPHklq19mFFeuoW+clHTiBfbeGZpSzed9eUIxpdz6j3D5ZmSg0QASkBWr7RVgHawAoBAUKjGRjK/sRk+IhKaRSnqRuXQ8o2m3n11hwdFO90GYr1FhIiPHW1fmPQxDDuwYvMUecQER4/PYmvX8zgleUC3nv6cEsvgDTqlnnkVAqnpxXv/P13idfTgZ5WPqgAKVNqYCLss1U+z+EX76gResL6gRjMixeRbcMJ+wRmvzTaIFKCvnYwVQzUsF9Zq6w5+omlUGuBMQjxaiN+D8I+NzaK/b/HvWpSZ4bHv/fMpPa5v/eMeJl0vyGNukWICL/8vjvwwLEE3nEi5cgaXGMcFCzNlBqYjIobdA2M7gopLj/bmBQiYuoRR2igVC3MsfuEZq4Xi/1eM8qao3Pj+SjFZFhM6t9ULID1Un/5hT+JOuGpA8C7b5+Ei5QY1T1zMUfW2E9ITd0G3/v2OXzv2+cc+//zNr6DPfW6sDQwvtaoBmKlesvWgAyOYU1dQPtXTtTvQbmp5ITbNcaiKnlNyS8NMRp3NODB8lZ16Ht4d1BR+vNk1I/MAE89oxp7JzR1QCl++v4H5rGQCArrorqfkUZ9HxPwuhEd0iogU2rg9qmokLWMeupFmwMyOLGAIoV0u2zohSYqMAsogVLGlG6HdsbQAfZnhWp78nlAZMxTL9Ra2vBzOxgZlME9dRGaOqAMj3l1Od/3tc1yE24XIelgAPPf/ej9jv2/9xtSftnnKAVIuw0tTwMT5anzi3e0p26/MyGg5C4zNtpD1UaqiSg+EtipUZTX7HIpla5Ggsb5aksbnmIHI/IL99RFGdqpqB/rxUbflFkeGzoKXvQ4kEZ9n5OO+vvm9xZqLbQ6TJhR97hdiAe9Bo26fWPGbyJGngx8HhcCXnsBSUBXeCPAqNsdK6fHSIuGbpchX20KMbJGBmVomrogoz4d86PW6vQNVG+WG8JbBBxlpFHf5wxqFZARWE3KmQj7DGnqIjx1HoAzEpgVESQFekZdRLBURC91jpG88VK9jS6DEE/dSO+dXLUJF4mRvQBgKjo4rdGpwqOjijTq+5x0xNc3+0VkNSknGfYZMrJCPHUDpeOAuL4vADQdXYT8IkpTB3gm0Cg5RJznbCQ4q0g94iSRYQVITrUIOKpIo77PSUf8fVsFZAQ28+IkQz5sVUYYWUGeczJkNDDbRlTgMGJAnPwi6gnCSDFQTmCKoZHc+HxVTOERhxcg7ZQSGWPYLDcdS2c8ikijvs+ZHDDWjl8cUzFxRj0V9mKrMmoohyD5JcTlFyOeujiPGLDvqfOe8mI99VHZKOJSDI30s89Vm0KkHg7/nu701Iu1NpodZ1oEHFWkUd/nDGoVkCk14Pe4bDeU0pMM+5CrDG7qpQ3IELBmLOCFi4DCSLlHvPxit6pUxLAKPeOWX4w8JeWqLaEphlG/BwGva1dVaaas5qhLTV0Y0qjvc9IDWgVslJR0RpE9ZybCPjQ7XVSa/btCiiwEcrkI8aB3tKdetz+flBMRZNR7fV9EBUq9Iwdl9FIM7a+ZMlBolq82hTa+IiJMxwK7AqUZdQiM9NTFIY36PkcbQN1HfhHt3Wge3ICLPas+LUxExFzsidDowKzIQKnf44LHRbazX0T1v+FEA56RgzLyajaKiHPBbwzZAY3ilPVawqcDKbnq2+WX3sBpadRFIY36PkfffldPxoGMgVEeXFbdg6jhwImQd2j2S6PdsT0PVQ8RKf1fbHvqoo26sRTDeNArJBuF1yQMuqHWWx3UWh3hk7ymYoFdgdIrG2UQAQuHeBD0uJFGfZ8T8LoRC3iwUtjt4Qj31EcY9V7GjZiLPTnCU+fGU4TcwxHRqVHT1P1i9hXTApeD9yVa454I+5Ad8DnzG21c4HkHFE99p/zyxmoRp9JhhHyyY4koDBl1IvoAEV0kostE9LE+rx8joq8R0UtE9AoRfUj8Vo8u98zF8Yqub0ar08VWpSncqI9q6qXJL2Py1EW2COBEBXRqFDX9iWOkU2NecDZKKuzD1gD5RWRQVs9UNIByo70t++jCahF3z8WFrnPUGWnUicgN4BMAPgjgbgAfJqK7d7ztXwP4PGPsAQBPAvj3ojd6lHnweAJvrJZQVUexcePqlKc+yHvOVhrwqAFOIeuFfFo5ej/yNfEeo4iRdtpYOeHyyxBPvSLWU08NqR7uGXWxnvq0mtbIvfVCrYXlXA13zx7+drjjxIin/jCAy4yxJcZYE8DnADyx4z0MAP9k4gBWxG1R8uCxJDpdhleWCwB0OeqCeqlzon4PPC4aeLFvlppICWy8lAx5UWl2BvYg2XSgFULYL8JTFx8o1f9/+1GotYRmo0xEfNgacEMVmROvR2sVoAZL31gtAgDumhXTaVSiYMSozwO4qfv7snpMz28B+AkiWgbwRQC/1O9/REQfIaJzRHQuk8lY2O7R5IFjSQDAC9dzAIA315SLYS4h1qgTEZJDPLhspYEJgcHZ3pT5URq+uDVFBUp9bhf8HvtNxoCevDSqGEik55wM+ZCrNPumUXKjLmpABod76uvqzfrCivI9vvsIDK4YJ6ICpR8G8MeMsQUAHwLwJ0S06//NGPsUY+wsY+zs5OThnxUoilTYh1PpMF66oRj1v3llFQvJoCOPrcOaemXKTS3FUgSjqkpFzmDlRAQESkuCespzeIrooGEojXYH1abYbJRU2Id2l2mTpfQ4qakDwEq+BkDR09MRv/AnzqOOEaN+C8Ci7u8L6jE9Pwfg8wDAGHsGQABAWsQGJQoPHEvixRt5ZEoN/LfLm3ji/jlHhl0Py0jJCm6ROqqyMVNqIBnywStgBitHxEi7ckNc210A8HvcSIS8Awcz9+QQcZ4zv5Fk+7SFyFeb8Atqd6wnFvTgzHQUT51fAWMMF1aK0kt3ACNXy/MA7iCik0TkgxIIfWrHe24AeB8AENFdUIy61FcE8uDxBLYqTXzy61fQ6TI8cf9OBUwMwwJo2XJTy5ARATdSA+UXB3LxlUBpB93u4OrNUYjqVKmnX2EOxwnPOaVmMPX7rEWnT3KICD/1zuO4sFrEM0tZXNooySCpA4w06oyxNoBfBPAlAG9AyXJ5nYh+m4i+T33bvwTw80T0MoDPAvgZNqzmWWKaB1Vd/TPPXMOdM1GcnnYmuJQM9y/drzTaqLU6Qiv/ep76APnFgVx83irATgZMWdAAaD1T0d0l9JxcxQFPfUj6quj0ST0/8MA8YgEP/tenLqDVYdJTdwBD30zG2BehBED1xz6u+/kCgMfEbk2i5/R0FBG/ogc75aUDQEqVXzpdBrcuy6WXoy5SUx/eUz1TauDEibCw9QD9SLuO5W6TxXoLi6mQyG1hKurH1c1K39dETyEChhea5R3y1AEg5PPgR88u4o++eRUApKfuALKi9IDgdhHuX0wAAP7ZfbOOrZMKK7NDC7XthtaJHh1Bnxt+j6uv/MIYc6S/Ta9T4+hBz4MoN9pCu2MCwGRMGVvY7wFX9LxQoHdz7ldVmqs2hWe+6PnJR4+DCAh4XTiZFnvTlhj01CX7g5991wk8dDyJhaRYL1GP3oPTZ53wvi9pQdWknESofw+SUqONRlt8n21ujI0Meh6EE5r6dDSAZqereMk7nob4+REpiQS8boR87oGeejzo3NCK4xNh/Hf3zqJUb297GpSIQRr1A8R33jmN77xz2tE1UgOqSrlHJ6pDI0fJttntNTsxgxXoGcbCiJa/g2CMCc9+AXpDJDZKjV1GPV9tIuh1C89G6RcUZ4whXxPfoXEn/8+TD0Cac2eQ8otkG/wRf2dbVp5DLdqoK/1fdnuLThl1o2P0BlFpdtDpMqH9aIBeDne/DJhcteVI4LJfTUKx3kanyxzT1DluFwmrTJZsRxp1yTa40e7nqUcDHmFVlJxxe+rcWA0bEDEM3mtefGHO9r4oekQPrOD0qx7OOyD1SMaLNOqSbaTCPhABq/1a/TownSYR8vXNftGMumhNPeCB20VDu0MOw4lCIEAvv/T31J2QQ/rJL3kHgrKS8SKNumQbfo8bx1MhXN4obTueLTeESy+A0iogX93dgyRTbsDrFtcRkuNyERJB78BmVqPQCoEED5AI+TyI+D27ZnjyNZ0wskpP9e3r8Zu5nBl6cJFGXbKLO6ajeGu9vO2YUk3qhKfuRbvLdvVjyZSUlgRO6K6DdHwjONWWFlC89Z2TgQDFe3ZCDkmF/ai3ulpLZwC4tK7czO+YjghfTzIepFGX7OL0dATXNitotHsDqDfLDaQFTTzSkxhQgOREjjonFfZpVZpmcaotLcAnA22XX7pdhryDnjqwPb5wcb2EY6mQnER0gJFGXbKL09NRtLtMq3Bsd7rIVVuOeOqDslE2HdLwAWMDrweh5YwLloWA/q0CSvU2usyZwGW/qtK31kuOtaCQjAdp1CW74Bc1l2D4RS+y7S5nUPtdLr84QcqGUc9XW4gFPPAI7BzJ4U299PEFXsnrRDwjtaOqtNnuYilTwZkZKb0cZKRRl+zi1GQYbhdp+upmmRt18UaWSyxrhZp2rNNlyDowg5WTUJuWWek5p5TQO5MZMhVTNO5SY7fGfdukeEOryS/q57u0WUa7y6SnfsCRRl2yC7/HjeMTIbylGhSeISFy6hFnMRlCxO/B6+oUHABaQzGnjHoy5EOz3UW12Rn95h0ohUAOGXVt3FtPgnlzrQQi4I4p8YY2Fdkuv1xcUz7vMzPSqB9kpFGX9OX0VC8DhmdkOCG/uFyEu+diePVWQTvmVOERJ2WjqjQveKycnn656hfXSjgxEUbQJ7boC1D64HjdpKV3vrVegsdFOJWW8stBRhp1SV9Oz0RxPVtBvdXB376yilTYh/lk0JG17p2P443VItodZQC100adBx2tZMBsVZzJRAF6nro+rfHieglnHJJDiAjziSBevplX1lor42Q6DJ9HmoWDjPz0JH05PR1BlwF/99oqvvrmBn7mnSeEtwjg3DsfR73VxeXM9icDp7JfkgOalhnBqZxxQOepq/JLrdnBtWzFUTnkR84u4ltXsri0XlIyX6T0cuCRRl3SFx4s+9//9g0EvW785CPHHVvrbfPKoIRXlxUJhmd8OKmpA+aNerPdRbnRdsxTj/o9CHhdWlOvSxslMAbcNeucof3ww8fg87jwyW9cwY2tqmNPBZLxIY26pC8nJsLwuAjZShM/9o5FxzI+AOBkOoKQz60FS59dymI2HtAGWohGS6M02dQrX3OumhRQ5BB9rvqbqzxw6dx0oFTYhyfum8NfvqjMkpeZLwcfadQlffF5XFpq48+966Sja7ldhHvUYOmtfA3feCuDH35owbH14kEviAbPRh2Ek9WknNunInj+2hbanS7eXCsh4HXhmODReTv56Xee0H6WmS8HH2nUJQP5iUeO41fed4fweZz9eNt8HBdWivjct28AAH707KJja3ncLsQC/ScuDYN79ikHn1qefMciVgt1fPnCOi6uF3F6Our4dKC3zcfx8InUWG4gEueRDR4kA/mpR0+Mba175+P4z61r+E/fvIp33Z52/EaSDHlNe+o5h9ru6nnfXdOYTwTxX565hssbZXzHmSnH1tLzf/7QvbiWrcjxcocA6alL9gX3zscBANVmBx9++Jjj6yXDPtOdGvNVZwZk6HG7CD/56HE8u7SFzXJzbHLIqcmI46MSJeNBGnXJvuDUZARBrxsTYR/ef5fzxiUZ2j0gYhS5MQ2Q+LGzi/CrueJ3zToXJJUcTqRRl+wL3C7Cz7/nFH79A2fGUvyi9FQ3Gyhtwu9xOVLdqScZ9uH77psDIAOXEvNITV2yb/jV7zo9trVSljx156pJd/IbH7oL333PjGOdKiWHF+mpS44kybAPtVYH9Zbxpl45B6tJd5IK+/Bdd0uNW2IeadQlR5LkgIlLw8g52PdFIhGFNOqSIwmvCjUjwSi91MfjqUskVpFGXXIk6c1GNW7U8w72UpdIRCGNuuRIwqtCtwwadcYY8rWW1otdItmvSKMuOZIMmo06iGK9jU6XjS1QKpFYRRp1yZFEk18MaurjqCaVSEQgjbrkSOLzuBAPerGuGx03DK2aVAZKJfscQ0adiD5ARBeJ6DIRfazP6/83EZ1X/7xFRHnhO5VIBHMiHca1zaqh9/KOjjJQKtnvjKwoJSI3gE8A+C4AywCeJ6KnGGMX+HsYY/+T7v2/BOABB/YqkQjlVDqMb1/dMvTejDpiLh2WFZ6S/Y0RT/1hAJcZY0uMsSaAzwF4Ysj7PwzgsyI2J5E4yYmJMG7la4aqSpc2K/C5XY4N35ZIRGHEqM8DuKn7+7J6bBdEdBzASQBftb81icRZTk6GAQDXspWR713KlHF8IiT7jUv2PaIDpU8C+AJjrK/rQ0QfIaJzRHQuk8kIXloiMceptGLUr2YMGPXNCk6pNwGJZD9jxKjfAqCfLbagHuvHkxgivTDGPsUYO8sYOzs5OWl8lxKJA5xQjfrS5nCj3u50cT1bwcl0ZBzbkkhsYcSoPw/gDiI6SUQ+KIb7qZ1vIqI7ASQBPCN2ixKJM0T8HkxF/bg6wqgv52podZj01CUHgpFGnTHWBvCLAL4E4A0An2eMvU5Ev01E36d765MAPscYY85sVSIRz8l0eKRRX9osAwBuk0ZdcgAwNCSDMfZFAF/ccezjO/7+W+K2JZGMh1OTYfzD6+tD37Okau6npPwiOQDIilLJkeZkOoxspYnCkB4wS5sVJEJeJMOy8Eiy/5FGXXKk4cHPq0PSGpcyZS1TRiLZ70ijLjnSnORpjapu3o+lTAWnJqX0IjkYSKMuOdIcS4XgosG56qV6Cxulhsx8kRwYpFGXHGl8HhcWU6GBueo8M0bKL5KDgjTqkiPPyXQYlzf6yy+aUZfyi+SAII265Mjz2G1pvLlWwqvLhV2vXclU4CLg+ERoD3YmkZhHGnXJkefJhxcR9Xvwh09f2fXaSzdyWEyF4Pe492BnEol5pFGXHHmiAS/++SPH8XevruK6LrXxuaUs/unSJn7sHYtD/rVEsr+QRl0iAfCzj52Ax+XCp55eAgAwxvA7f/cmZmIB/OxjJ/d4dxKJcQy1CZBIDjtTsQB+8MF5/PkLyzgzE0XA68b5m3n8mx96OwJeKb1IDg7SqEskKr/8vjtwYbWIj//16wCA09MR/NBDC3u8K4nEHNKoSyQqc4kg/voXHsPLywX81Uu38P0PzMtJR5IDhzTqEokOIsL9iwncv5jY661IJJaQgVKJRCI5REijLpFIJIcIadQlEonkECGNukQikRwipFGXSCSSQ4Q06hKJRHKIkEZdIpFIDhHSqEskEskhghhje7MwUQbAdQv/NA1gU/B2RLFf97Zf9wXIvVlhv+4LkHuzipm9HWeMTQ56cc+MulWI6Bxj7Oxe76Mf+3Vv+3VfgNybFfbrvgC5N6uI3JuUXyQSieQQIY26RCKRHCIOolH/1F5vYAj7dW/7dV+A3JsV9uu+ALk3qwjb24HT1CUSiUQymIPoqUskEolkANKoSyQSySFiXxl1IvoRInqdiLpEdHbHa79BRJeJ6CIRfY/u+AfUY5eJ6GO64yeJ6Dn1+J8RkU/gPv+MiM6rf64R0Xn1+Akiqule+0Pdv3mIiF5V9/P7ROTISB0i+i0iuqXbw4d0r5k6h4L39X8R0ZtE9AoR/VciSqjH9/yc9dmr4+djxPqLRPQ1IrqgXg//o3rc9Gfr0P6uqZ/LeSI6px5LEdGXieiS+t+kepzUz+6y+tk/6NCezujOy3kiKhLRr+zVOSOiTxPRBhG9pjtm+hwR0U+r779ERD9taHHG2L75A+AuAGcAfB3AWd3xuwG8DMAP4CSAKwDc6p8rAE4B8KnvuVv9N58H8KT68x8C+O8d2vPvAfi4+vMJAK8NeN+3ATwCgAD8HYAPOrSf3wLwa32Omz6Hgvf13QA86s+/C+B398s527HmWM7HiD3MAnhQ/TkK4C318zP12Tq4v2sA0juO/RsAH1N//pju8/2Q+tmR+lk+N6bPcA3A8b06ZwDeA+BB/Xfb7DkCkAKwpP43qf6cHLX2vvLUGWNvMMYu9nnpCQCfY4w1GGNXAVwG8LD65zJjbIkx1gTwOQBPqB7ddwL4gvrv/wuA7xe9X3WdHwXw2RHvmwUQY4w9y5RP6zNO7GcEps6h6MUZY//AGGurf30WwNCJznt4zsZyPobBGFtljL2o/lwC8AaA+SH/ZNBnO06egHKdAduvtycAfIYpPAsgoX62TvI+AFcYY8Mq1h09Z4yxpwFs9VnTzDn6HgBfZoxtMcZyAL4M4AOj1t5XRn0I8wBu6v6+rB4bdHwCQF5nRPhx0bwbwDpj7JLu2EkieomIvkFE79btf7nPPp3iF9XHuE/zRzyYP4dO8rNQPBPOfjhnnL04HwMhohMAHgDwnHrIzGfrFAzAPxDRC0T0EfXYNGNsVf15DcD0Hu0NAJ7EdkdrP5wzwPw5srTHsRt1IvpHInqtz5+xekOjMLjPD2P7l2cVwDHG2AMAfhXA/0dEsTHv7ZMAbgNwv7qf3xO9vsV98ff8JoA2gD9VD43lnB1EiCgC4C8A/ApjrIg9/Gx38C7G2IMAPgjgF4joPfoX1SerPcmVJiV29n0A/lw9tF/O2TacPEceJ/6nw2CMvd/CP7sFYFH39wX1GAYcz0J5hPGo3rr+/UL2SUQeAD8I4CHdv2kAaKg/v0BEVwCcVtfWyw2m92Nmb7o9/kcAf6v+1ew5FL4vIvoZAN8L4H3ql3ps58wEw87T2CAiLxSD/qeMsb8EAMbYuu51o5+tcBhjt9T/bhDRf4UiW6wT0SxjbFWVDjb2Ym9QbjQv8nO1X86ZitlzdAvAe3cc//qoRQ6K/PIUgCeJyE9EJwHcASWI9jyAO0jJdPFBeex6SjUYXwPww+q//2kAfy14T+8H8CZjTJMIiGiSiNzqz6fUfS6pj1xFInpE1eF/yoH98D3o9cofAMCj76bOoQP7+gCAXwfwfYyxqu74np+zHYzlfAxD/X3/E4A3GGP/Tnfc7GfrxN7CRBTlP0MJgL+m7oFnZ+ivt6cA/JSa4fEIgIJOgnCCbU/P++Gc6TB7jr4E4LuJKKnKRt+tHhuOqGiviD9QTvoyFM9tHcCXdK/9JpQI9UXosiCgRI7fUl/7Td3xU1A+pMtQHsX8gvf6xwA+uuPYDwF4HcB5AC8C+Ge6185C+UJdAfAHUKt5HTiHfwLgVQCvqF+WWavnUPC+LkPRB8+rf/5wv5yzPnt1/HyMWP9dUB7NX9Gdrw9Z+Wwd2NspKFkjL6uf22+qxycAfAXAJQD/CCClHicAn1D39ip0WW0O7C0M5Sk9rju2J+cMyo1lFUALik37OSvnCEr86bL6518YWVu2CZBIJJJDxEGRXyQSiURiAGnUJRKJ5BAhjbpEIpEcIqRRl0gkkkOENOoSiURyiJBGXSKRSA4R0qhLJBLJIeL/Bw9o1AEL2jNcAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "##plot for testing\n",
    "current = final_data[(final_data['endpoint']=='midpoint') & (final_data['GC_correction']=='GC_corrected')][plot_columns].mean()\n",
    "plt.plot(plot_columns,current)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #run the analysis \n",
    "\n",
    "# #start the timer\n",
    "# start_time=time.time()\n",
    "\n",
    "# p = Pool(processes=CPU) #use the specified number of processes\n",
    "# p.map(run_full_analysis, to_do_list, 1) #run the analysis on all TFs in TFs_list. Send only one item to each processor at a time.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print('merging all sites')\n",
    "# start_time = time.time()\n",
    "\n",
    "# # merge results together and export\n",
    "# merged_out_file = results_dir+'/coverage/all_sites/'+sample_name+'.all_sites.coverage.txt'\n",
    "# try: #make any necessary directorys for output\n",
    "#     os.mkdir(results_dir+'/coverage/all_sites')\n",
    "# except:\n",
    "#     pass \n",
    "# merged_output=pd.DataFrame()\n",
    "# for j,line in enumerate(to_do_list):\n",
    "#     if j%50 == 0:\n",
    "#         print(line[0],time.time()-start_time)\n",
    "#         sys.stdout.flush()\n",
    "#     site_name = line[0]\n",
    "#     indiv_out_file = results_dir+'/coverage/'+site_name+'/'+sample_name+'.'+site_name+'.coverage.txt'\n",
    "#     current_data = pd.read_csv(indiv_out_file,sep='\\t')\n",
    "#     merged_output = merged_output.append(current_data,ignore_index=True,sort=True)\n",
    "\n",
    "# merged_output.columns\n",
    "# merged_output.to_csv(merged_out_file,sep='\\t',index=False)\n",
    "        \n",
    "# print('done with merge')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
